<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>spammR lipidomic and proteomic integration • spammR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="spammR lipidomic and proteomic integration">
<meta name="description" content="A basic walkthrough of how to use the spammR package to perform a
spatial omics analysis for lipidomic and proteomic data.
">
<meta property="og:description" content="A basic walkthrough of how to use the spammR package to perform a
spatial omics analysis for lipidomic and proteomic data.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spammR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.17</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/lipidProt.html">spammR lipidomic and proteomic integration</a></li>
    <li><a class="dropdown-item" href="../articles/spatMicrobiome.html">spammR microbiome evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/spatProt.html">spammR Spatial Proteomics Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>spammR lipidomic and proteomic integration</h1>
                        <h4 data-toc-skip class="author">Sara
Gosline</h4>
            
            <h4 data-toc-skip class="date">Feb 04, 2026</h4>
      

      <div class="d-none name"><code>lipidProt.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="about">About<a class="anchor" aria-label="anchor" href="#about"></a>
</h2>
<p>The goal of this vignette is to showcase integration between two
different omes in the same spatial context. For this we use a dataset
from sections of rat brain measured via mass spetrometry imaging (MSI)
to identify lipidomic measurements alongside proteomics from regions of
interest from <a href="https://pubs.acs.org/doi/10.1021/acs.analchem.4c04462" class="external-link">Vandergrift
et al</a>.</p>
<p>We show how <code>spammR</code> enables: - download and loading of
datasets - visualization of features within each dataset -
identification of correlated regions in this dataset</p>
<p>First we load the required packages.</p>
</div>
<div class="section level2">
<h2 id="load-proteomics-data">Load proteomics data<a class="anchor" aria-label="anchor" href="#load-proteomics-data"></a>
</h2>
<p>The proteomics data can be found on <a href="https://zenodo.org/records/13345212" class="external-link">zenodo</a> and downloaded
directly. From there.</p>
<div class="section level3">
<h3 id="format-data">Format data<a class="anchor" aria-label="anchor" href="#format-data"></a>
</h3>
<p>We require formatting the data as a matrix for loading.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#brain data</span></span>
<span><span class="va">finame</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'https://zenodo.org/records/13345212/files/DESI+spatial'</span>,</span>
<span>                 <span class="st">'%20proteomics%20rat%20brain%20proteomics%20'</span>,</span>
<span>                 <span class="st">'result.xlsx?download=1'</span><span class="op">)</span></span>
<span><span class="va">fi</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="va">finame</span>,</span>
<span>                    dest <span class="op">=</span> <span class="st">'dat.xlsx'</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#read in data, convert to matrix </span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">readxl</span><span class="fu">::</span><span class="fu"><a href="https://readxl.tidyverse.org/reference/read_excel.html" class="external-link">read_xlsx</a></span><span class="op">(</span><span class="st">'dat.xlsx'</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">feature_data</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Protein</span>,<span class="va">`Protein ID`</span>,<span class="va">`Entry Name`</span>,<span class="va">Gene</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/distinct.html" class="external-link">distinct</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">'Protein'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dat</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">dat</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">Protein</span>, <span class="fu">starts_with</span><span class="op">(</span><span class="st">'ROI'</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">'Protein'</span><span class="op">)</span></span></code></pre></div>
<p>The image itself we downloaded separately from metaspace (see below)
and stored it in the package.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##read in image</span></span>
<span><span class="va">img</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,<span class="st">"brain_img_0.png"</span>,package <span class="op">=</span> <span class="st">'spammR'</span><span class="op">)</span></span></code></pre></div>
<p>Now that the image is loaded we can visualize it (not done here to
save space) to verify it is what we expect.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html" class="external-link">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_image.html" class="external-link">draw_image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>                                                    <span class="st">"brain_img_0.png"</span>,</span>
<span>                                                    package <span class="op">=</span> <span class="st">"spammR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>As you can see, there are squares chopped out of the image. These
were the samples that were measured via proteomics. The next step is to
map these coordinates so we can compare them with the MSI data.</p>
</div>
<div class="section level3">
<h3 id="loading-image-coordinates-rom-geojson">Loading image coordinates rom GEOJson<a class="anchor" aria-label="anchor" href="#loading-image-coordinates-rom-geojson"></a>
</h3>
<p>We used the program [QuPath][(<a href="https://qupath.readthedocs.io/en/stable/" class="external-link uri">https://qupath.readthedocs.io/en/stable/</a>) to collect the
regions of interest (ROI) coordinates on the image. This program allowed
us to highlight the regions of interest and export them in GEOJson
format, which can then be used as input into spammR.</p>
<p>We store the GEOJson file in the package for you to load with this
example should you choose to load teo GEOJson using the
<code>geojsonR</code> package.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#remove this once we can include it in spammR</span></span>
<span> </span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mlampros/geojsonR" class="external-link">geojsonR</a></span><span class="op">)</span></span>
<span>  <span class="va">jsondat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/geojsonR/man/FROM_GeoJson.html" class="external-link">FROM_GeoJson</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,<span class="st">'brain_roi.geojson'</span>,</span>
<span>                                      package <span class="op">=</span> <span class="st">'spammR'</span><span class="op">)</span><span class="op">)</span></span>
<span> </span>
<span>  <span class="co">##herr er hry yhr</span></span>
<span>  <span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">jsondat</span><span class="op">$</span><span class="va">features</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>     <span class="va">roi</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">properties</span><span class="op">$</span><span class="va">name</span></span>
<span>     <span class="va">xv</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">geometry</span><span class="op">$</span><span class="va">coordinates</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span></span>
<span>     <span class="va">yv</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">$</span><span class="va">geometry</span><span class="op">$</span><span class="va">coordinates</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span></span>
<span>     <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">roi</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="va">roi</span> <span class="op">=</span> <span class="st">""</span></span>
<span>     <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>ID <span class="op">=</span> <span class="va">roi</span>, x_pixels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">xv</span><span class="op">)</span>, y_pixels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">yv</span><span class="op">)</span>,</span>
<span>                  spot_height <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">yv</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">yv</span><span class="op">)</span>,</span>
<span>                 spot_width <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">max</a></span><span class="op">(</span><span class="va">xv</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html" class="external-link">min</a></span><span class="op">(</span><span class="va">xv</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>     <span class="op">}</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ID</span> <span class="op">!=</span> <span class="st">""</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu">tidyr</span><span class="fu">::</span><span class="fu"><a href="https://tidyr.tidyverse.org/reference/separate.html" class="external-link">separate</a></span><span class="op">(</span><span class="va">ID</span>,into <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'ROI'</span>,<span class="st">'Replicate'</span><span class="op">)</span>, </span>
<span>                     sep <span class="op">=</span> <span class="st">'_'</span>, </span>
<span>                     remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">remove_rownames</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>     <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">column_to_rownames</a></span><span class="op">(</span><span class="st">'ID'</span><span class="op">)</span></span>
<span> </span>
<span> <span class="co">##now for each ROI we want an x, y, cell height and cell_width</span></span>
<span> <span class="co">#y-coordinates are from top, so need to udpate</span></span>
<span>  <span class="va">coords</span><span class="op">$</span><span class="va">y_pixels</span> <span class="op">=</span> <span class="fl">263</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">coords</span><span class="op">$</span><span class="va">y_pixels</span><span class="op">)</span> <span class="op">-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">coords</span><span class="op">$</span><span class="va">spot_height</span><span class="op">)</span></span>
<span>  <span class="va">coords</span><span class="op">$</span><span class="va">x_pixels</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">coords</span><span class="op">$</span><span class="va">x_pixels</span><span class="op">)</span></span>
<span>  <span class="va">coords</span><span class="op">$</span><span class="va">spot_height</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">coords</span><span class="op">$</span><span class="va">spot_height</span><span class="op">)</span></span>
<span>  <span class="va">coords</span><span class="op">$</span><span class="va">spot_width</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">coords</span><span class="op">$</span><span class="va">spot_width</span><span class="op">)</span></span>
<span>  <span class="fu">readr</span><span class="fu">::</span><span class="fu"><a href="https://readr.tidyverse.org/reference/write_delim.html" class="external-link">write_csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span>,file<span class="op">=</span><span class="st">'../inst/extdata/bcoords.csv'</span><span class="op">)</span></span></code></pre></div>
<p>To avoid this we just load the csv directly:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>,<span class="st">'bcoords.csv'</span>,</span>
<span>                               package <span class="op">=</span> <span class="st">'spammR'</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">coords</span><span class="op">$</span><span class="va">ROI</span>, <span class="va">coords</span><span class="op">$</span><span class="va">Replicate</span>,sep <span class="op">=</span> <span class="st">'_'</span><span class="op">)</span></span></code></pre></div>
<p>Now that we have the coordinates we can use spammR to create the
<code>SpatialExperiment</code> object and plot a protein.</p>
</div>
<div class="section level3">
<h3 id="create-spatialexperiment-object">Create SpatialExperiment Object<a class="anchor" aria-label="anchor" href="#create-spatialexperiment-object"></a>
</h3>
<p>Here we load the proteomics and create the object.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span>   <span class="co">#create an SFE</span></span>
<span>  <span class="va">spe</span> <span class="op">&lt;-</span> <span class="fu">spammR</span><span class="fu">::</span><span class="fu"><a href="../reference/convert_to_spe.html">convert_to_spe</a></span><span class="op">(</span>dat <span class="op">=</span> <span class="va">dat</span>, </span>
<span>                      feature_meta <span class="op">=</span> <span class="va">feature_data</span>,</span>
<span>                      sample_meta <span class="op">=</span> <span class="va">coords</span>,</span>
<span>                      spatial_coords_colnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'x_pixels'</span>,<span class="st">'y_pixels'</span><span class="op">)</span>,</span>
<span>                      assay_name <span class="op">=</span> <span class="st">'proteomics'</span>,</span>
<span>                      sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>, </span>
<span>                      image_files <span class="op">=</span> <span class="va">img</span>,</span>
<span>                      image_id <span class="op">=</span> <span class="st">'rat_brain'</span><span class="op">)</span></span>
<span></span>
<span>  <span class="co"># myelin</span></span>
<span>  <span class="fu">spammR</span><span class="fu">::</span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">spe</span>, feature <span class="op">=</span> <span class="st">'Mbp'</span>, </span>
<span>                          feature_type <span class="op">=</span> <span class="st">'Gene'</span>,</span>
<span>                sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>, </span>
<span>                image_id <span class="op">=</span> <span class="st">'rat_brain'</span><span class="op">)</span></span></code></pre></div>
<p><img src="lipidProt_files/figure-html/create%20proteomics%20spe-1.png" class="r-plt" alt="" width="700"></p>
<p>The protein data was collected from smaller regions of interest
within the larger image.</p>
</div>
</div>
<div class="section level2">
<h2 id="load-lipidomics-data">Load lipidomics data<a class="anchor" aria-label="anchor" href="#load-lipidomics-data"></a>
</h2>
<p>Now that we have the image and protein data loaded, we can begin to
ingest the lipid data from metaspace. This requires an additional
command to pull the information directly.</p>
<div class="section level3">
<h3 id="retrieve-metaspace-data">Retrieve metaspace data<a class="anchor" aria-label="anchor" href="#retrieve-metaspace-data"></a>
</h3>
<p>We created a wrapper to the <a href="">Metaspace python package</a>
to enable download and formatting of metaspace measurements for
spammR.</p>
<p>The result is a <code>SpatialExperiment</code>. This takes too long
to run as a vignette so we show the code and then load the file
separately.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##get lipid data from metaspace </span></span>
<span></span>
<span><span class="va">mspe</span> <span class="op">&lt;-</span> <span class="fu">spammR</span><span class="fu">::</span><span class="fu"><a href="../reference/retrieve_metaspace_data.html">retrieve_metaspace_data</a></span><span class="op">(</span><span class="st">"2024-02-15_20h37m13s"</span>, </span>
<span>                                fdr <span class="op">=</span> <span class="fl">0.2</span>, </span>
<span>                                assay_name <span class="op">=</span> <span class="st">'lipids'</span>,</span>
<span>                                sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>,</span>
<span>                                rotate <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                                drop_zeroes <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now we can downlod the object from figshare and load it.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">'https://api.figshare.com/v2/file/download/60993349'</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">'mspe.rds'</span><span class="op">)</span></span>
<span><span class="va">mspe</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="st">'mspe.rds'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">'mspe.rds'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Lastly we add the image file to the data downloaded by metaspace and
visualize a lipid of interest.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## add in image to check </span></span>
<span><span class="va">mspe</span> <span class="op">&lt;-</span> <span class="fu">SpatialExperiment</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SpatialExperiment/man/imgData-methods.html" class="external-link">addImg</a></span><span class="op">(</span><span class="va">mspe</span>, <span class="va">img</span>, scaleFactor <span class="op">=</span> <span class="cn">NA_real_</span>,</span>
<span>                                  sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>,</span>
<span>                                  image_id <span class="op">=</span> <span class="st">'rat_brain'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">mspe</span><span class="op">)</span> </span>
<span><span class="va">meta</span><span class="op">$</span><span class="va">Name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">vapply</a></span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">moleculeNames</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">}</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">mspe</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">meta</span></span>
<span></span>
<span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">mspe</span>, feature <span class="op">=</span> <span class="st">'C41H78NO7P-H-'</span>,</span>
<span>                 assay <span class="op">=</span> <span class="st">'lipids'</span>,</span>
<span>                 plot_log <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                   metric_display <span class="op">=</span> <span class="st">'Lipid expression'</span>,</span>
<span>                sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">'rat_brain'</span><span class="op">)</span><span class="co">#plot with an ion</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 17673 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_label()`).</span></span></code></pre>
<p><img src="lipidProt_files/figure-html/add%20image-1.png" class="r-plt" alt="" width="700"></p>
<p>Here we can see the overlay of specific lipids.</p>
</div>
<div class="section level3">
<h3 id="merging-images-to-single-coordinate-system">Merging images to single coordinate system<a class="anchor" aria-label="anchor" href="#merging-images-to-single-coordinate-system"></a>
</h3>
<p>Since the proteomics is measured in discrete ROI and the lipids are
measured on a per-pixel basis, to carry out pure multiomic comparisons
we need to map them to the same coordinate space.</p>
<p>To do this we use the <code>spat_reduce</code> function to reduce the
lipidomics data to the same regions as the proteomics.</p>
<p><img src="lipidProt_files/figure-html/reduce-1.png" class="r-plt" alt="" width="700"></p>
<p>Now we can start to analyze the relationship between lipids and
proteins. # Correlation analysis</p>
<p>One of the innovative analyses we can do with the spatial data is to
evaluate correlations within molecules across space. We include the
<code>spatial_network</code> function that can compute correlations
between groups of features across omic domains and show how to use it
below.</p>
</div>
<div class="section level3">
<h3 id="ascribing-function-to-unknown-lipids">Ascribing function to unknown lipids<a class="anchor" aria-label="anchor" href="#ascribing-function-to-unknown-lipids"></a>
</h3>
<p>We can use the correlation analysis to assess the biological function
of specific lipids, by seeking out which pathways are over-represented
among proteins who are correlated with a lipid of interest.</p>
<div class="section level4">
<h4 id="protein-correlations">Protein correlations<a class="anchor" aria-label="anchor" href="#protein-correlations"></a>
</h4>
<p>Let’s see which of our lipids are most variable across the
tissue.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lvs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">mspe</span>,<span class="st">'lipids'</span><span class="op">)</span>,<span class="fl">1</span>,<span class="va">var</span>,na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></span>
<span></span>
<span><span class="va">lvn</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">lvs</span><span class="op">)</span>,<span class="st">'Name'</span><span class="op">]</span></span>
<span></span>
<span><span class="va">lvn</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "PI(18:1(11Z)/20:3(8Z,11Z,14Z))"</span></span></code></pre>
<p>We’ll focus on phosphatidylserin
“PS(28:6(10Z,13Z,16Z,19Z,22Z,25Z)/12:0)”, which is the most variable
across the sample.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#number 1 seems to have an interesting pattern</span></span>
<span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">mspe</span>,feature <span class="op">=</span> <span class="va">lvn</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, assay_name <span class="op">=</span> <span class="st">'lipids'</span>, </span>
<span>                feature_type <span class="op">=</span> <span class="st">'Name'</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">'rat_brain'</span>, sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>, </span>
<span>                metric_display <span class="op">=</span> <span class="st">'Lipid expression'</span>,</span>
<span>                plot_log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 68794 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_label()`).</span></span></code></pre>
<p><img src="lipidProt_files/figure-html/plot%20lipid%20of%20interest-1.png" class="r-plt" alt="" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">reduced</span>,feature <span class="op">=</span> <span class="va">lvn</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, assay_name <span class="op">=</span> <span class="st">'lipids'</span>, </span>
<span>                feature_type <span class="op">=</span> <span class="st">'Name'</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">'rat_brain'</span>, sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>, </span>
<span>                metric_display <span class="op">=</span> <span class="st">'Lipid expression'</span>,</span>
<span>                plot_log <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 25 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_label()`).</span></span></code></pre>
<p><img src="lipidProt_files/figure-html/plot%20lipid%20of%20interest-2.png" class="r-plt" alt="" width="700"></p>
<p>Now we can ask what proteins correlate with this lipid in the reduced
object.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#let's look at the correlation between the lipid and proteins</span></span>
<span><span class="va">fl</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">lvn</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span><span class="op">[</span>,<span class="st">'Gene'</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">cg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_network.html">spatial_network</a></span><span class="op">(</span><span class="va">reduced</span>, </span>
<span>                      assay_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"proteomics"</span>,<span class="st">"lipids"</span><span class="op">)</span>, </span>
<span>                      query_features <span class="op">=</span> <span class="va">lvn</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="co">#lipid</span></span>
<span>                      target_features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span><span class="op">[</span>,<span class="st">'Gene'</span><span class="op">]</span>,</span>
<span>                      feature_names <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'Gene'</span>,<span class="st">'Name'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(rowval)`</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##we can then use trick above to get the neighborhood of our lipid</span></span>
<span><span class="va">gt</span> <span class="op">&lt;-</span> <span class="va">cg</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">activate</span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">neigh_weights</span> <span class="op">&lt;-</span>  <span class="va">cg</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">activate</span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">from</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">gt</span><span class="op">$</span><span class="va">name</span> <span class="op">==</span> <span class="va">lvn</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="co">#then we can get the edge weights and nodes </span></span>
<span><span class="va">edges</span> <span class="op">&lt;-</span> <span class="va">neigh_weights</span> <span class="op">|&gt;</span> <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">nodes</span> <span class="op">&lt;-</span> <span class="va">neigh_weights</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">activate</span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">as_tibble</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##create a table with the protein correlationm</span></span>
<span><span class="va">prot_cor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">nodes</span><span class="op">[</span><span class="va">edges</span><span class="op">$</span><span class="va">to</span>,<span class="op">]</span>,lipid_cor <span class="op">=</span> <span class="va">edges</span><span class="op">$</span><span class="va">corval</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span>Gene <span class="op">=</span> <span class="st">'name'</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">Gene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span><span class="op">[</span>,<span class="st">'Gene'</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="va">neigh_weights</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">activate</span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">corval</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0.75</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">activate</span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>degree <span class="op">=</span> <span class="fu">centrality_degree</span><span class="op">(</span>mode <span class="op">=</span> <span class="st">'all'</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">degree</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">ggraph</span><span class="op">(</span>layout <span class="op">=</span> <span class="st">'stress'</span><span class="op">)</span>  <span class="op">+</span></span>
<span>  <span class="fu">geom_edge_link</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>colour <span class="op">=</span> <span class="va">corval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu">geom_node_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">class</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu">geom_node_label</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>label <span class="op">=</span> <span class="va">name</span>, color <span class="op">=</span> <span class="va">class</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="lipidProt_files/figure-html/correlation%20network-1.png" class="r-plt" alt="" width="700"></p>
<p>Now we have the correlation value - both negative and positive - that
we can use as input to a functional enrichment test.</p>
</div>
<div class="section level4">
<h4 id="functional-enrichment">Functional enrichment<a class="anchor" aria-label="anchor" href="#functional-enrichment"></a>
</h4>
<p>We can look at the pathway enriched by these correlation values using
leapR via the <code>enrichment_in_order</code> method.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##add back to data set</span></span>
<span></span>
<span><span class="va">prot_cor</span> <span class="op">&lt;-</span> <span class="va">prot_cor</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu">summarize</span><span class="op">(</span>lipid_cor <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">lipid_cor</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">left_join</span><span class="op">(</span><span class="va">prot_cor</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>upperGene <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">toupper</a></span><span class="op">(</span><span class="va">Gene</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(Gene)`</span></span></code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"krbpaths"</span><span class="op">)</span></span>
<span><span class="va">enrich</span> <span class="op">&lt;-</span> <span class="fu">leapR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/leapR/man/leapR.html" class="external-link">leapR</a></span><span class="op">(</span>geneset <span class="op">=</span> <span class="va">krbpaths</span>, enrichment_method <span class="op">=</span> <span class="st">'enrichment_in_order'</span>,</span>
<span>                       eset <span class="op">=</span> <span class="va">reduced</span>, primary_column <span class="op">=</span> <span class="st">'lipid_cor'</span>, </span>
<span>                       id_column <span class="op">=</span> <span class="st">'upperGene'</span>, minsize <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">enrich</span>, <span class="va">BH_pvalue</span> <span class="op">&lt;</span> <span class="fl">0.3</span><span class="op">)</span></span>
<span><span class="co">#subset(enrich,ingroup_mean &gt; 0.25) |&gt; </span></span>
<span><span class="co">#  subset(pvalue &lt;0.05)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                 ingroup_n</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM        20</span></span>
<span><span class="co">##                                                                                                                                                            ingroupnames</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM GOT2, ASS1, GLUL, GLUD1, GOT1, GLS, GAD1, ASL, GPT, PPAT, ASNS, ABAT, GFPT1, GAD2, ACY3, ASPA, ALDH5A1, ADSL, NIT2, CAD</span></span>
<span><span class="co">##                                                 ingroup_mean outgroup_n</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM    0.2539142       5040</span></span>
<span><span class="co">##                                                 outgroup_mean   zscore</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM   0.003796164 0.728826</span></span>
<span><span class="co">##                                                 oddsratio       pvalue</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM 0.3849837 7.634525e-05</span></span>
<span><span class="co">##                                                  BH_pvalue SignedBH_pvalue</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM 0.04481466      0.04481466</span></span>
<span><span class="co">##                                                 background_n background_mean</span></span>
<span><span class="co">## KEGG_ALANINE_ASPARTATE_AND_GLUTAMATE_METABOLISM           NA              NA</span></span></code></pre>
<p>We can now look at pathways that are the most statistically
significant.</p>
</div>
<div class="section level4">
<h4 id="plot-proteins">Plot proteins<a class="anchor" aria-label="anchor" href="#plot-proteins"></a>
</h4>
<p>We first check the expression of the proteins as well and see if they
are all correlated in the same direction.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu">stringr</span><span class="fu">::</span><span class="fu"><a href="https://stringr.tidyverse.org/reference/str_split.html" class="external-link">str_split</a></span><span class="op">(</span><span class="va">paths</span><span class="op">[</span><span class="fl">1</span>,<span class="st">'ingroupnames'</span><span class="op">]</span>, pattern <span class="op">=</span> <span class="st">', '</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">reduced</span>,<span class="va">upperGene</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">prots</span><span class="op">)</span>,<span class="st">'proteomics'</span><span class="op">)</span></span>
<span><span class="va">lv</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/altExps.html" class="external-link">altExp</a></span><span class="op">(</span><span class="va">reduced</span><span class="op">)</span>,<span class="va">Name</span> <span class="op">==</span> <span class="va">lvn</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,<span class="st">'lipids'</span><span class="op">)</span></span>
<span> </span>
<span><span class="co">##proteins seem anti-correlated</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">pv</span><span class="op">)</span>,<span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">lv</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">'spearman'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] -0.2047532</span></span></code></pre>
<p>We can see that they are negative correlated.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">reduced</span>, assay_name <span class="op">=</span> <span class="st">'proteomics'</span>,</span>
<span>                feature <span class="op">=</span> <span class="va">prots</span>,</span>
<span>                feature_type <span class="op">=</span> <span class="st">'upperGene'</span>,</span>
<span>                sample_id <span class="op">=</span> <span class="st">'rat_brain'</span>,</span>
<span>                image_id <span class="op">=</span> <span class="st">'rat_brain'</span>, </span>
<span>                plot_log <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>                title_size <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                plot_title <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">paths</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,</span>
<span>                metric_display <span class="op">=</span> <span class="st">'Protein expression'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">p2</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 25 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_label()`).</span></span></code></pre>
<p><img src="lipidProt_files/figure-html/plot%20proteins-1.png" class="r-plt" alt="" width="700"></p>
<p>The heatmap confirms that the lipid of interest is indeed associated
with this glucose and suger transporter pathway.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>This vignette explores how one can use the spammR package to analyze
multiomics measurements captured in a spatial context.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sara Gosline.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
