<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>spammR Spatial Proteomics Example • spammR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="spammR Spatial Proteomics Example">
<meta name="description" content="A basic walkthrough of how to use and analyze spatial proteomics
measurements in the spammR package.
">
<meta property="og:description" content="A basic walkthrough of how to use and analyze spatial proteomics
measurements in the spammR package.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spammR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.17</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/lipidProt.html">spammR lipidomic and proteomic integration</a></li>
    <li><a class="dropdown-item" href="../articles/spatMicrobiome.html">spammR microbiome evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/spatProt.html">spammR Spatial Proteomics Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>spammR Spatial Proteomics Example</h1>
                        <h4 data-toc-skip class="author">Sara
Gosline</h4>
            
            <h4 data-toc-skip class="date">Feb 04, 2026</h4>
      

      <div class="d-none name"><code>spatProt.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>To install the package currently you must install directly from
<code>GitHub</code> along with the leapR dependency as shown below.
Before release we hope to move to Bioconductor.</p>
<p>The <code>leapR</code> package is designed for flexible pathway
enrichment and currently must be installed before spammR.</p>
<pre><code><span>  <span class="co">##install if not already installed</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'PNNL-CompBio/leapR'</span><span class="op">)</span></span>
<span>  <span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">'PNNL-CompBio/spammR'</span><span class="op">)</span></span></code></pre>
<p>Once the package is installed you load the library, including the
test data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load spammR</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spammR</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="collecting-data-to-analyze">Collecting data to analyze<a class="anchor" aria-label="anchor" href="#collecting-data-to-analyze"></a>
</h2>
<p><code>spammR</code> enables the analysis of disparate sets of
multiomic data: image-based data and numerical measurements of omics
data. It is incredibly flexible as to the <em>type</em> of multiomic
data. We assume each omics measurement is collected in a single sample,
and that there are specific spatial coordinates for that sample in the
image. We leverage the <code>SpatialExperiment</code> object to store
the data for each image/measurement pair.</p>
<div class="section level3">
<h3 id="data-overview-and-examples">Data overview and examples<a class="anchor" aria-label="anchor" href="#data-overview-and-examples"></a>
</h3>
<p>The <code>spammR</code> package requires omics data with spatial
coordinates for the functions to run successfully. Here we describe the
data required and show examples.</p>
<div class="section level4">
<h4 id="omics-measurement-data">Omics Measurement Data<a class="anchor" aria-label="anchor" href="#omics-measurement-data"></a>
</h4>
<p><code>SpatialExperiment</code> can hold multiple omics measurements
mapping to the same sample identifier in different ‘slots’. This data
can be a tabular data frame or matrix with rownames referencing
measurements in a particular sample (e.g. gene, species) and column
names representing sample identifiers. An example of this can be found
by loading <code>pancDataList.rda</code> file from Figshare.</p>
<p>To evaluate the features of this package we are using pancreatic data
from <a href="">Gosline et al.</a> that is captured using mass
spectrometry measured from 7 independent regions of a single human
pancreas. Each image is segmented into nine ‘voxels’, with one voxel per
image representing a cluster of islet cells.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://api.figshare.com/v2/file/download/55158821"</span>,</span>
<span>              mode <span class="op">=</span> <span class="st">"wb"</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span>, dest <span class="op">=</span> <span class="st">"pdl.rda"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"pdl.rda"</span><span class="op">)</span></span>
<span><span class="fu">utils</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pancDataList</span><span class="op">$</span><span class="va">Image_0</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            0_S_1_1  0_S_1_2  0_S_1_3  0_S_2_1  0_S_2_2  0_S_2_3</span></span>
<span><span class="co">## sp|A0A024RBG1|NUD4B_HUMAN 13.06042 13.42317 12.42396 13.02470 12.56442 12.69023</span></span>
<span><span class="co">## sp|A0A096LP55|QCR6L_HUMAN 15.10920 15.27460 15.16780 15.01030 15.46639 14.73712</span></span>
<span><span class="co">## sp|A0AV96|RBM47_HUMAN     17.40246 17.29727 17.25559 17.34851 17.12866 17.17658</span></span>
<span><span class="co">## sp|A0AVT1|UBA6_HUMAN      18.00653 18.43015 18.24663 18.17563 18.38961 18.29268</span></span>
<span><span class="co">## sp|A0FGR8|ESYT2_HUMAN     16.59018 16.48890 16.50134 16.55334 16.32316 16.43397</span></span>
<span><span class="co">## sp|A0MZ66|SHOT1_HUMAN     18.19277 18.73633 18.54485 18.20005 18.74041 18.70762</span></span>
<span><span class="co">##                            0_S_3_1  0_S_3_2</span></span>
<span><span class="co">## sp|A0A024RBG1|NUD4B_HUMAN       NA 13.40670</span></span>
<span><span class="co">## sp|A0A096LP55|QCR6L_HUMAN 14.81792 15.63741</span></span>
<span><span class="co">## sp|A0AV96|RBM47_HUMAN     17.27792 17.11678</span></span>
<span><span class="co">## sp|A0AVT1|UBA6_HUMAN      18.12080 18.10220</span></span>
<span><span class="co">## sp|A0FGR8|ESYT2_HUMAN     16.43682 16.20515</span></span>
<span><span class="co">## sp|A0MZ66|SHOT1_HUMAN     18.60198 18.62946</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"pdl.rda"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Here the rownames represent protein identifiers and the column names
represent individual samples. Each element of the list contains the
measurements from a different sample:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">pancDataList</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] 7</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pancDataList</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                            1_S_1_1  1_S_1_2  1_S_1_3  1_S_2_1  1_S_2_2  1_S_2_3</span></span>
<span><span class="co">## sp|A0A024RBG1|NUD4B_HUMAN       NA       NA       NA       NA       NA       NA</span></span>
<span><span class="co">## sp|A0A096LP55|QCR6L_HUMAN       NA       NA       NA       NA       NA       NA</span></span>
<span><span class="co">## sp|A0AV96|RBM47_HUMAN     17.68866 17.58076 17.51335 17.65900 17.52926 17.44999</span></span>
<span><span class="co">## sp|A0AVT1|UBA6_HUMAN      18.02235 18.35493 18.03651 18.09462 18.08796 18.04204</span></span>
<span><span class="co">## sp|A0FGR8|ESYT2_HUMAN     17.50177 17.51525 17.34338 17.41622 17.34254 17.46062</span></span>
<span><span class="co">## sp|A0MZ66|SHOT1_HUMAN     18.57445 18.67262 18.83672 18.69094 18.48785 18.75190</span></span>
<span><span class="co">##                            1_S_3_1  1_S_3_2</span></span>
<span><span class="co">## sp|A0A024RBG1|NUD4B_HUMAN       NA       NA</span></span>
<span><span class="co">## sp|A0A096LP55|QCR6L_HUMAN       NA       NA</span></span>
<span><span class="co">## sp|A0AV96|RBM47_HUMAN     17.37103 17.50367</span></span>
<span><span class="co">## sp|A0AVT1|UBA6_HUMAN      18.21052 17.94707</span></span>
<span><span class="co">## sp|A0FGR8|ESYT2_HUMAN     17.31495 17.53217</span></span>
<span><span class="co">## sp|A0MZ66|SHOT1_HUMAN     18.82805 18.48856</span></span></code></pre>
<p>This list is used below in our analysis examples.</p>
</div>
<div class="section level4">
<h4 id="sample-metadata">Sample metadata<a class="anchor" aria-label="anchor" href="#sample-metadata"></a>
</h4>
<p>The samples metadata table contains mappings between samples and
metadata. An example can be found in <code>data(pancMeta)</code>. Most
importantly we require the image mapping information, which includes: -
<em>Image coordinates:</em> to map the image to a coordinate space we
need to know the <code>x_origin</code>, and <code>y_origin</code>
(assumed to be zero) as well as <code>x_max</code> and
<code>y_max</code>, which is the top right of the image. The package
plots the <em>entire</em> image so specifying these coordinates ensures
that all other points are properly mapped. - <em>Sample
coordinates:</em> Each sample has its own <code>x_coord</code> and
<code>y_coord</code>. - <em>Spot size:</em> <code>spot_height</code> and
<code>spot_width</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pancMeta</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">pancMeta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##         Image x_coord y_coord IsletStatus IsletOrNot Plex Grid.Number x_pixels</span></span>
<span><span class="co">## 0_S_3_1     0       3       1    Proximal   NonIslet 127N           1      475</span></span>
<span><span class="co">## 0_S_2_1     0       2       1       Islet      Islet 128N           2      380</span></span>
<span><span class="co">## 0_S_1_1     0       1       1    Proximal   NonIslet 127C           3      285</span></span>
<span><span class="co">## 0_S_3_2     0       3       2    Proximal   NonIslet 128C           4      475</span></span>
<span><span class="co">## 0_S_2_2     0       2       2    Proximal   NonIslet 129N           5      380</span></span>
<span><span class="co">## 0_S_1_2     0       1       2    Proximal   NonIslet 129C           6      285</span></span>
<span><span class="co">##         y_pixels x_origin y_origin x_max y_max spot_width spot_height</span></span>
<span><span class="co">## 0_S_3_1      170        0        0   860   725         90         140</span></span>
<span><span class="co">## 0_S_2_1      170        0        0   860   725         90         140</span></span>
<span><span class="co">## 0_S_1_1      170        0        0   860   725         90         140</span></span>
<span><span class="co">## 0_S_3_2      315        0        0   860   725         90         140</span></span>
<span><span class="co">## 0_S_2_2      315        0        0   860   725         90         140</span></span>
<span><span class="co">## 0_S_1_2      315        0        0   860   725         90         140</span></span></code></pre>
<p>This metadata contains information for all 7 images, so we do not
need a separate metadata file for each image, the
<code>convert_to_spe</code> function will simply take the metadata
relevant to the data file.</p>
</div>
<div class="section level4">
<h4 id="image-files">Image files<a class="anchor" aria-label="anchor" href="#image-files"></a>
</h4>
<p>There can be multiple image files associated with a single set of
omics measurements. Currently we have tested working with files in
<code>png</code> format. Each image we have is stained so that we can
identify the Islet cells. Each image also has a grid superimposed to
show where the sample measurements came from. The grid is not necessary,
of course, but can help alibrate the coordinates.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/ggdraw.html" class="external-link">ggdraw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu">cowplot</span><span class="fu">::</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/draw_image.html" class="external-link">draw_image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>,</span>
<span>                                                    <span class="st">"Image_1.png"</span>,</span>
<span>                                                    package <span class="op">=</span> <span class="st">"spammR"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Now we can use this image and others to visualize omics data.</p>
</div>
<div class="section level4">
<h4 id="omics-metadata">Omics metadata<a class="anchor" aria-label="anchor" href="#omics-metadata"></a>
</h4>
<p>The last set of metadata relates to the <code>rows</code> of the
omics measurement data. When using gene-based data, this will be the
genes or proteins in the dataset. When using metagenomics, this will
refer to the species. One column of this table must uniquely map to the
rownames of the omics data.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">protMeta</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">protMeta</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"pancProts"</span>, <span class="st">"EntryName"</span>, <span class="st">"PrimaryGeneName"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                   pancProts   EntryName PrimaryGeneName</span></span>
<span><span class="co">## 1 sp|A0A024RBG1|NUD4B_HUMAN NUD4B_HUMAN          NUDT4B</span></span>
<span><span class="co">## 2 sp|A0A096LP55|QCR6L_HUMAN QCR6L_HUMAN          UQCRHL</span></span>
<span><span class="co">## 3     sp|A0AV96|RBM47_HUMAN RBM47_HUMAN           RBM47</span></span>
<span><span class="co">## 4      sp|A0AVT1|UBA6_HUMAN  UBA6_HUMAN            UBA6</span></span>
<span><span class="co">## 5     sp|A0FGR8|ESYT2_HUMAN ESYT2_HUMAN           ESYT2</span></span>
<span><span class="co">## 6     sp|A0MZ66|SHOT1_HUMAN SHOT1_HUMAN           SHTN1</span></span></code></pre>
<p>This data helps us find better gene identifiers.</p>
</div>
</div>
<div class="section level3">
<h3 id="loading-data-into-spatial-experiment-object-">Loading data into spatial experiment object.<a class="anchor" aria-label="anchor" href="#loading-data-into-spatial-experiment-object-"></a>
</h3>
<p>Now that we have all the data loaded we can build a
<code>SpatialExperiment</code> object either using ALL samples or just
the samples in a single image. We can pool all the data for more
statistical power.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pooledData</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="va">pancDataList</span><span class="op">)</span></span>
<span><span class="va">pooled.panc.spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/convert_to_spe.html">convert_to_spe</a></span><span class="op">(</span><span class="va">pooledData</span>, <span class="co">## pooled data table</span></span>
<span>  <span class="va">pancMeta</span>, <span class="co">## pooled metadata</span></span>
<span>  <span class="va">protMeta</span>, <span class="co">## protein identifiers</span></span>
<span>  feature_meta_colname <span class="op">=</span> <span class="st">"pancProts"</span>, <span class="co"># column name</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Spatial object created without spatial coordinate </span></span>
<span><span class="co">##          column names provided. Distance based analysis will not be enabled.</span></span></code></pre>
<pre><code><span><span class="co">## Note: Only mapping metadata for 6662 features out of 6693 data points</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pooled.panc.spe</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: SpatialExperiment </span></span>
<span><span class="co">## dim: 6662 63 </span></span>
<span><span class="co">## metadata(0):</span></span>
<span><span class="co">## assays(1): proteomics</span></span>
<span><span class="co">## rownames(6662): sp|A0A024RBG1|NUD4B_HUMAN sp|A0A096LP55|QCR6L_HUMAN ...</span></span>
<span><span class="co">##   sp|Q9Y3M8|STA13_HUMAN sp|Q9Y6X3|SCC4_HUMAN</span></span>
<span><span class="co">## rowData names(6): pancProts Entry ... GeneNames PrimaryGeneName</span></span>
<span><span class="co">## colnames(63): 0_S_1_1 0_S_1_2 ... 3_S_3_2 3_S_3_3</span></span>
<span><span class="co">## colData names(16): Image x_coord ... spot_height sample_id</span></span>
<span><span class="co">## reducedDimNames(0):</span></span>
<span><span class="co">## mainExpName: NULL</span></span>
<span><span class="co">## altExpNames(0):</span></span>
<span><span class="co">## spatialCoords names(0) :</span></span>
<span><span class="co">## imgData names(0):</span></span></code></pre>
<p>We can also create a list of <code>SpatialExperiment</code> objects,
one for each of the 3 images we have.</p>
<p>Now we can use these individual image objects or the combined
‘pooled’ object for analysis.</p>
</div>
<div class="section level3">
<h3 id="spatial-data-with-image">Spatial data with image<a class="anchor" aria-label="anchor" href="#spatial-data-with-image"></a>
</h3>
<p>Here we loop over all of the images in <code>imglist</code> to plot
the expression of the insulin protein in each image. We expect insulin
(or INS) to be highest in voxels containing islet cells, which we label
using the <code>label_column</code> ‘IsletOrNot’ which was loaded into
the metadata for us.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">allimgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">imglist</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">img.spes</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span></span>
<span>    <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>      feature <span class="op">=</span> <span class="st">"INS"</span>,</span>
<span>      feature_type <span class="op">=</span> <span class="st">"PrimaryGeneName"</span>,</span>
<span>      sample_id <span class="op">=</span> <span class="va">x</span>,</span>
<span>      image_id <span class="op">=</span> <span class="st">"with_grid"</span>,</span>
<span>      label_column <span class="op">=</span> <span class="st">"IsletOrNot"</span>,</span>
<span>      interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>  <span class="op">)</span></span>
<span>  <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span></span>
<span><span class="va">allimgs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span></code></pre></div>
<p><img src="spatProt_files/figure-html/spatial%20data-1.png" class="r-plt" alt="" width="700"></p>
<p>To go further and visualize entire pathways we need to first identify
which groups of proteins are of interest using a more unsupervised
approach.</p>
</div>
</div>
<div class="section level2">
<h2 id="expression-and-pathway-analysis">Expression and pathway analysis<a class="anchor" aria-label="anchor" href="#expression-and-pathway-analysis"></a>
</h2>
<p>Now that we have the ability to overlay omic measurements with image
ones, we can identify new features to plot and visualize them. First we
can employ standard differential expression approaches using the voxel
labels and the <code>limma</code> pathway.</p>
<div class="section level3">
<h3 id="differential-expression">Differential expression<a class="anchor" aria-label="anchor" href="#differential-expression"></a>
</h3>
<p>First we want to identify specific proteins that are up-regulated in
the islet cells (or regions labeled ‘islet’) compared to other regions.
We can then plot the set of proteins.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">islet_res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calc_spatial_diff_ex.html">calc_spatial_diff_ex</a></span><span class="op">(</span><span class="va">pooled.panc.spe</span>,</span>
<span>    assay_name <span class="op">=</span> <span class="st">"proteomics"</span>,</span>
<span>    log_transformed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>    category_col <span class="op">=</span> <span class="st">"IsletOrNot"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># we filter the significant proteins first</span></span>
<span><span class="va">sig_prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">islet_res</span><span class="op">)</span>, </span>
<span>                    <span class="va">NonIslet_vs_Islet.adj.P.Val.limma</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="co"># then separate into up-regulated and down-regulated based on fold chnage</span></span>
<span><span class="va">ups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sig_prots</span>, <span class="va">NonIslet_vs_Islet.logFC.limma</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">downs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sig_prots</span>, <span class="va">NonIslet_vs_Islet.logFC.limma</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>  <span class="st">"We found"</span>, <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sig_prots</span><span class="op">)</span>, <span class="st">"significantly differentally \</span></span>
<span><span class="st">  expressed proteins including"</span>,</span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ups</span><span class="op">)</span>, <span class="st">"upregulated proteins and"</span>, <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">downs</span><span class="op">)</span>, <span class="st">"downregulated"</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We found 241 significantly differentally \n  expressed proteins including 168 upregulated proteins and 73 downregulated"</span></span></code></pre>
<p>Now we can plot those differentially expressed proteins across
images.</p>
</div>
<div class="section level3">
<h3 id="plot-differentially-expressed-proteins">Plot differentially expressed proteins<a class="anchor" aria-label="anchor" href="#plot-differentially-expressed-proteins"></a>
</h3>
<p>If we are interested in the combined expression of proteins we can
also visualize those.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">spe.plot</span> <span class="op">&lt;-</span> <span class="va">img.spes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">hup</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">spe.plot</span>,</span>
<span>    feature <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">downs</span><span class="op">)</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="st">"Image_1"</span>,</span>
<span>    image_id <span class="op">=</span> <span class="st">"with_grid"</span>,</span>
<span>    label_column <span class="op">=</span> <span class="st">"IsletOrNot"</span>,</span>
<span>    interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">hup</span></span></code></pre></div>
<p><img src="spatProt_files/figure-html/plot%20pathway-1.png" class="r-plt" alt="" width="700"></p>
</div>
<div class="section level3">
<h3 id="pathway-enrichment-measurements">Pathway enrichment measurements<a class="anchor" aria-label="anchor" href="#pathway-enrichment-measurements"></a>
</h3>
<p>Now we can calculate the enriched pathways in the islets.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">leapR</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"krbpaths"</span><span class="op">)</span></span>
<span><span class="va">ora.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_ora.html">enrich_ora</a></span><span class="op">(</span><span class="va">islet_res</span>, geneset <span class="op">=</span> <span class="va">krbpaths</span>, </span>
<span>                      geneset_name <span class="op">=</span> <span class="st">"krbpaths"</span>, </span>
<span>                      feature_column <span class="op">=</span> <span class="st">"PrimaryGeneName"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">ora.res</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"INSULIN"</span>, <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">ora.res</span><span class="op">)</span><span class="op">)</span>, </span>
<span>              <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ingroup_n"</span>, <span class="st">"pvalue"</span>, <span class="st">"BH_pvalue"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                                                                                                           ingroup_n</span></span>
<span><span class="co">## KEGG_INSULIN_SIGNALING_PATHWAY                                                                                    5</span></span>
<span><span class="co">## BIOCARTA_INSULIN_PATHWAY                                                                                          2</span></span>
<span><span class="co">## REACTOME_GLUCOSE_REGULATION_OF_INSULIN_SECRETION                                                                 13</span></span>
<span><span class="co">## REACTOME_INSULIN_SYNTHESIS_AND_SECRETION                                                                         24</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_ACTIVITY_BY_INSULIN_LIKE_GROWTH_FACTOR_BINDING_PROTEINS         0</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION                                                                         16</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_ACETYLCHOLINE                                                         7</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_GLUCAGON_LIKE_PEPTIDE_1                                              11</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_FREE_FATTY_ACIDS                                                      7</span></span>
<span><span class="co">## REACTOME_INHIBITION_OF_INSULIN_SECRETION_BY_ADRENALINE_NORADRENALINE                                              4</span></span>
<span><span class="co">##                                                                                                                 pvalue</span></span>
<span><span class="co">## KEGG_INSULIN_SIGNALING_PATHWAY                                                                            1.000000e+00</span></span>
<span><span class="co">## BIOCARTA_INSULIN_PATHWAY                                                                                  2.247236e-01</span></span>
<span><span class="co">## REACTOME_GLUCOSE_REGULATION_OF_INSULIN_SECRETION                                                          5.549704e-02</span></span>
<span><span class="co">## REACTOME_INSULIN_SYNTHESIS_AND_SECRETION                                                                  2.522014e-07</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_ACTIVITY_BY_INSULIN_LIKE_GROWTH_FACTOR_BINDING_PROTEINS 1.000000e+00</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION                                                                  3.897737e-02</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_ACETYLCHOLINE                                                 1.077909e-04</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_GLUCAGON_LIKE_PEPTIDE_1                                       9.083131e-06</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_FREE_FATTY_ACIDS                                              2.897170e-05</span></span>
<span><span class="co">## REACTOME_INHIBITION_OF_INSULIN_SECRETION_BY_ADRENALINE_NORADRENALINE                                      1.628147e-02</span></span>
<span><span class="co">##                                                                                                              BH_pvalue</span></span>
<span><span class="co">## KEGG_INSULIN_SIGNALING_PATHWAY                                                                            1.0000000000</span></span>
<span><span class="co">## BIOCARTA_INSULIN_PATHWAY                                                                                  0.9987657450</span></span>
<span><span class="co">## REACTOME_GLUCOSE_REGULATION_OF_INSULIN_SECRETION                                                          0.7062802680</span></span>
<span><span class="co">## REACTOME_INSULIN_SYNTHESIS_AND_SECRETION                                                                  0.0002100838</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_LIKE_GROWTH_FACTOR_ACTIVITY_BY_INSULIN_LIKE_GROWTH_FACTOR_BINDING_PROTEINS 1.0000000000</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION                                                                  0.5797883996</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_ACETYLCHOLINE                                                 0.0128271210</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_GLUCAGON_LIKE_PEPTIDE_1                                       0.0037831242</span></span>
<span><span class="co">## REACTOME_REGULATION_OF_INSULIN_SECRETION_BY_FREE_FATTY_ACIDS                                              0.0070612576</span></span>
<span><span class="co">## REACTOME_INHIBITION_OF_INSULIN_SECRETION_BY_ADRENALINE_NORADRENALINE                                      0.3686245089</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="pathway-plotting">Pathway plotting<a class="anchor" aria-label="anchor" href="#pathway-plotting"></a>
</h3>
<p>We know that there are significantly enriched pathways in insulin
secretion, so let’s plot those.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">secprots</span> <span class="op">&lt;-</span> <span class="va">ora.res</span><span class="op">[</span><span class="st">"REACTOME_GLUCOSE_REGULATION_OF_INSULIN_SECRETION"</span>, <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ingroupnames</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span>split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe.plot</span> <span class="op">&lt;-</span> <span class="va">img.spes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">hup</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">spe.plot</span>,</span>
<span>    feature <span class="op">=</span> <span class="va">secprots</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="st">"Image_1"</span>,</span>
<span>    image_id <span class="op">=</span> <span class="st">"with_grid"</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"PrimaryGeneName"</span>,</span>
<span>    label_column <span class="op">=</span> <span class="st">"IsletOrNot"</span>,</span>
<span>    plot_title <span class="op">=</span> <span class="st">"Glucose regulation proteins"</span>,</span>
<span>    interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">hup</span></span></code></pre></div>
<p><img src="spatProt_files/figure-html/pathway%20plotting-1.png" class="r-plt" alt="" width="700"></p>
<p>The average expression of the 20 proteins selected is shown to be
higher in islet cells than adjacent cells.</p>
</div>
</div>
<div class="section level2">
<h2 id="distance-based-measurements">Distance based measurements<a class="anchor" aria-label="anchor" href="#distance-based-measurements"></a>
</h2>
<p>We can also identify features that are correlated with distance to a
feature or a gradient in the sample. This will provide input to
rank-based statistical tools that can help identify pathways.</p>
<div class="section level3">
<h3 id="distance-based-measurements-1">Distance based measurements<a class="anchor" aria-label="anchor" href="#distance-based-measurements-1"></a>
</h3>
<p>First we identify a specific feature, the Islet cell, and use that to
identify proteins correlated with distance from the islet in each image.
Proteins with a negative correlation are decreasing in expression as
they are farther from the islet cells.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## for each image, let's compute the distance of each voxel to </span></span>
<span><span class="co">## the one labeled 'Islet'</span></span>
<span><span class="va">rank.imgs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span></span>
<span>  <span class="va">img.spes</span>,</span>
<span>  <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="../reference/distance_based_analysis.html">distance_based_analysis</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"proteomics"</span>,</span>
<span>      sampleCategoryCol <span class="op">=</span> <span class="st">"IsletOrNot"</span>,</span>
<span>      sampleCategoryValue <span class="op">=</span> <span class="st">"Islet"</span></span>
<span>    <span class="op">)</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co">## now we have the distances, let's plot some interesting proteins</span></span>
<span><span class="va">negProts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rank.imgs</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rank.imgs</span><span class="op">[[</span><span class="va">x</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span>, </span>
<span>           <span class="va">IsletDistancespearmanPval</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">IsletDistancespearmanCor</span> <span class="op">&lt;</span> <span class="op">(</span><span class="op">-</span><span class="fl">.75</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">PrimaryGeneName</span>, <span class="va">IsletDistancespearmanCor</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>      <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>image <span class="op">=</span> <span class="va">x</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">negProts</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##                       PrimaryGeneName IsletDistancespearmanCor   image</span></span>
<span><span class="co">## sp|A6NFH5|FBP12_HUMAN          FABP12               -0.9621024 Image_0</span></span>
<span><span class="co">## sp|O14979|HNRDL_HUMAN         HNRNPDL               -0.9114654 Image_0</span></span>
<span><span class="co">## sp|O15145|ARPC3_HUMAN           ARPC3               -0.8861469 Image_0</span></span>
<span><span class="co">## sp|O43292|GPAA1_HUMAN           GPAA1               -0.8355100 Image_0</span></span>
<span><span class="co">## sp|O43829|ZBT14_HUMAN          ZBTB14               -0.8608285 Image_0</span></span>
<span><span class="co">## sp|O60256|KPRB_HUMAN          PRPSAP2               -0.8355100 Image_0</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## do any proteins show up more than once?</span></span>
<span><span class="va">icounts</span> <span class="op">&lt;-</span> <span class="va">negProts</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">PrimaryGeneName</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>numImgs <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">numImgs</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">icounts</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 267 × 2</span></span></span>
<span><span class="co">##    PrimaryGeneName numImgs</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>             <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> CHGA                  2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> DCAKD                 2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> ECHS1                 2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> PFKL                  2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> PKM                   2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> PSMD4                 2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> SH3GL1                2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> VAMP2                 2</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> AAK1                  1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> ABCB7                 1</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 257 more rows</span></span></span></code></pre>
<p>It looks like SH3GL1 is correlated with distance to Islet in a few
images.</p>
</div>
<div class="section level3">
<h3 id="plot-protein-gradient">Plot protein gradient<a class="anchor" aria-label="anchor" href="#plot-protein-gradient"></a>
</h3>
<p>Now we can plot the expression of a protein suspected to have
decreasing expression farther from the islet cells.We start with SH3GL1
and SP3S2.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    feature <span class="op">=</span> <span class="st">"SH3GL1"</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"PrimaryGeneName"</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">)</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span>,</span>
<span>    image_id <span class="op">=</span> <span class="st">"with_grid"</span>,</span>
<span>    label_column <span class="op">=</span> <span class="st">"IsletOrNot"</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spatProt_files/figure-html/plot%20correlated%20proteins-1.png" class="r-plt" alt="" width="700"></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    feature <span class="op">=</span> <span class="st">"AP3S2"</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"PrimaryGeneName"</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>    image_id <span class="op">=</span> <span class="st">"with_grid"</span>,</span>
<span>    label_column <span class="op">=</span> <span class="st">"IsletOrNot"</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spatProt_files/figure-html/plot%20correlated%20proteins-2.png" class="r-plt" alt="" width="700"></p>
<p>The expression of this protein is lower farther from the Islet. Can
we identify trends in the proteins?</p>
</div>
<div class="section level3">
<h3 id="gradient-based-enrichment">Gradient-based enrichment<a class="anchor" aria-label="anchor" href="#gradient-based-enrichment"></a>
</h3>
<p>Rank-based pathway enrichment is a way to evaluate trends pathways
that are over-represented in a ranked list of genes. The
<code>leapR</code> pathway has such functionality and we can use the
rankings as input.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">leapR</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"krbpaths"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">spe</span> <span class="op">&lt;-</span> <span class="va">rank.imgs</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">enriched.paths</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_gradient.html">enrich_gradient</a></span><span class="op">(</span><span class="va">spe</span>,</span>
<span>    geneset <span class="op">=</span> <span class="va">krbpaths</span>,</span>
<span>    feature_column <span class="op">=</span> <span class="st">"PrimaryGeneName"</span>, <span class="co"># mapped to enrichment data</span></span>
<span>    ranking_column <span class="op">=</span> <span class="st">"IsletDistancespearmanCor"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">enriched.paths</span><span class="op">[</span>, <span class="st">"comp"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">rank.imgs</span><span class="op">)</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">enriched.paths</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">enriched.paths</span><span class="op">[</span>, <span class="st">"krbpaths"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">enriched.paths</span><span class="op">)</span></span>
<span></span>
<span><span class="va">enriched.paths</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pvalue</span> <span class="op">&lt;</span> <span class="fl">0.01</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html" class="external-link">group_by</a></span><span class="op">(</span><span class="va">krbpaths</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html" class="external-link">summarize</a></span><span class="op">(</span>numImgs <span class="op">=</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/context.html" class="external-link">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/desc.html" class="external-link">desc</a></span><span class="op">(</span><span class="va">numImgs</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 2 × 2</span></span></span>
<span><span class="co">##   krbpaths                                 numImgs</span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> REACTOME_DIABETES_PATHWAYS                     1</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> REACTOME_INSULIN_SYNTHESIS_AND_SECRETION       1</span></span></code></pre>
<p>We can see that numerous pathways are coming up as enriched across
images, including ribosomal and translation related pathways. Now we can
select proteins from a particular pathway and visualize those as
well.</p>
</div>
<div class="section level3">
<h3 id="plotting-pathways-from-gradient">Plotting pathways from gradient<a class="anchor" aria-label="anchor" href="#plotting-pathways-from-gradient"></a>
</h3>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rprots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">enriched.paths</span>, <span class="va">krbpaths</span> <span class="op">==</span> <span class="st">"REACTOME_INSULIN_SYNTHESIS_AND_SECRETION"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">comp</span>, <span class="va">ingroupnames</span><span class="op">)</span></span>
<span></span>
<span><span class="va">rprots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">rprots</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">]</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>,</span>
<span>    feature <span class="op">=</span> <span class="va">rprots</span>,</span>
<span>    feature_type <span class="op">=</span> <span class="st">"PrimaryGeneName"</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>,</span>
<span>    image_id <span class="op">=</span> <span class="st">"with_grid"</span>,</span>
<span>    label_column <span class="op">=</span> <span class="st">"IsletOrNot"</span>, interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="spatProt_files/figure-html/gradient%20plotting-1.png" class="r-plt" alt="" width="700"></p>
<p>This shows the ribosomal protein expression across the image.</p>
</div>
<div class="section level3">
<h3 id="network-plotting">Network plotting<a class="anchor" aria-label="anchor" href="#network-plotting"></a>
</h3>
<p>We can also look at the correlation of the ribosomal proteins in a
graph. The correlation code takes a while but then we can reduce the
graph to the proteins we are most interested in, or those that are most
correlated.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidygraph.data-imaginist.com" class="external-link">tidygraph</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'tidygraph'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     active, slice</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     active, rename</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggraph.data-imaginist.com" class="external-link">ggraph</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##correlation analysis can be slow, so let's only evaluate the top 1000 most variable proteins</span></span>
<span><span class="va">varprots</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span>,<span class="fl">1</span>,<span class="va">var</span>,na.rm <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/sort.html" class="external-link">sort</a></span><span class="op">(</span>decreasing <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/date.html" class="external-link">date</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Wed Feb  4 15:21:54 2026"</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">full_graph</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/spatial_network.html">spatial_network</a></span><span class="op">(</span><span class="va">img.spes</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span><span class="op">[</span><span class="va">varprots</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">1000</span><span class="op">]</span>,<span class="op">]</span>,</span>
<span>                              <span class="st">'proteomics'</span>,<span class="st">'PrimaryGeneName'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Joining with `by = join_by(rowval)`</span></span></code></pre>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/date.html" class="external-link">date</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "Wed Feb  4 15:22:03 2026"</span></span></code></pre>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##how subset for only those 81 proteins</span></span>
<span><span class="va">rgraph</span> <span class="op">&lt;-</span> <span class="va">full_graph</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">nodes</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">name</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="va">rprots</span><span class="op">)</span> <span class="op">|&gt;</span><span class="co">#[sample(20)]) |&gt;</span></span>
<span>  <span class="fu">tidygraph</span><span class="fu">::</span><span class="fu"><a href="https://tidygraph.data-imaginist.com/reference/activate.html" class="external-link">activate</a></span><span class="op">(</span><span class="va">edges</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">corval</span> <span class="op">&gt;</span> <span class="fl">0.75</span><span class="op">)</span></span>
<span></span>
<span><span class="co">##then we can plot</span></span>
<span><span class="fu">ggraph</span><span class="fu">::</span><span class="fu"><a href="https://ggraph.data-imaginist.com/reference/ggraph.html" class="external-link">ggraph</a></span><span class="op">(</span><span class="va">rgraph</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_edge_link.html" class="external-link">geom_edge_link</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">corval</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_point.html" class="external-link">geom_node_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>   <span class="fu"><a href="https://ggraph.data-imaginist.com/reference/geom_node_text.html" class="external-link">geom_node_label</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="va">name</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Using <span style="color: #0000BB;">"stress"</span> as default layout</span></span></code></pre>
<p><img src="spatProt_files/figure-html/graph%20correlations-1.png" class="r-plt" alt="" width="700">
Here are the highly correlated edges between 20 randomly sampled
ribosomal proteins.</p>
</div>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>This vignette shows various functions to apply in managing spatial
proteomics data in spammR.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ol style="list-style-type: decimal">
<li>Gosline et al.</li>
<li>Spatial Experiment</li>
</ol>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Tahoe 26.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] ggraph_2.2.2                ggplot2_4.0.1              </span></span>
<span><span class="co">##  [3] tidygraph_1.3.1             leapR_0.99.6               </span></span>
<span><span class="co">##  [5] spammR_0.99.17              limma_3.66.0               </span></span>
<span><span class="co">##  [7] SpatialExperiment_1.20.0    SingleCellExperiment_1.32.0</span></span>
<span><span class="co">##  [9] SummarizedExperiment_1.40.0 Biobase_2.70.0             </span></span>
<span><span class="co">## [11] GenomicRanges_1.62.1        Seqinfo_1.0.0              </span></span>
<span><span class="co">## [13] IRanges_2.44.0              S4Vectors_0.48.0           </span></span>
<span><span class="co">## [15] BiocGenerics_0.56.0         generics_0.1.4             </span></span>
<span><span class="co">## [17] MatrixGenerics_1.22.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [19] BiocStyle_2.38.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] DBI_1.2.3           deldir_2.0-4        gridExtra_2.3      </span></span>
<span><span class="co">##   [4] s2_1.1.9            rlang_1.1.7         magrittr_2.0.4     </span></span>
<span><span class="co">##   [7] otel_0.2.0          e1071_1.7-17        compiler_4.5.2     </span></span>
<span><span class="co">##  [10] png_0.1-8           systemfonts_1.3.1   vctrs_0.7.1        </span></span>
<span><span class="co">##  [13] wk_0.9.5            pkgconfig_2.0.3     fastmap_1.2.0      </span></span>
<span><span class="co">##  [16] backports_1.5.0     magick_2.9.0        XVector_0.50.0     </span></span>
<span><span class="co">##  [19] labeling_0.4.3      utf8_1.2.6          rmarkdown_2.30     </span></span>
<span><span class="co">##  [22] tzdb_0.5.0          ragg_1.5.0          purrr_1.2.1        </span></span>
<span><span class="co">##  [25] xfun_0.56           cachem_1.1.0        jsonlite_2.0.0     </span></span>
<span><span class="co">##  [28] DelayedArray_0.36.0 tweenr_2.0.3        broom_1.0.12       </span></span>
<span><span class="co">##  [31] R6_2.6.1            bslib_0.10.0        RColorBrewer_1.1-3 </span></span>
<span><span class="co">##  [34] reticulate_1.44.1   boot_1.3-32         car_3.1-3          </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4     Rcpp_1.1.1          bookdown_0.46      </span></span>
<span><span class="co">##  [40] knitr_1.51          readr_2.1.6         Matrix_1.7-4       </span></span>
<span><span class="co">##  [43] igraph_2.2.1        tidyselect_1.2.1    rstudioapi_0.18.0  </span></span>
<span><span class="co">##  [46] abind_1.4-8         yaml_2.3.12         viridis_0.6.5      </span></span>
<span><span class="co">##  [49] lattice_0.22-7      tibble_3.3.1        withr_3.0.2        </span></span>
<span><span class="co">##  [52] S7_0.2.1            evaluate_1.0.5      sf_1.0-24          </span></span>
<span><span class="co">##  [55] desc_1.4.3          units_1.0-0         proxy_0.4-29       </span></span>
<span><span class="co">##  [58] spData_2.3.4        polyclip_1.10-7     pillar_1.11.1      </span></span>
<span><span class="co">##  [61] BiocManager_1.30.27 ggpubr_0.6.2        carData_3.0-5      </span></span>
<span><span class="co">##  [64] KernSmooth_2.23-26  plotly_4.12.0       sp_2.2-0           </span></span>
<span><span class="co">##  [67] hms_1.1.4           scales_1.4.0        class_7.3-23       </span></span>
<span><span class="co">##  [70] glue_1.8.0          lazyeval_0.2.2      tools_4.5.2        </span></span>
<span><span class="co">##  [73] data.table_1.18.2.1 ggnewscale_0.5.2    ggsignif_0.6.4     </span></span>
<span><span class="co">##  [76] fs_1.6.6            graphlayouts_1.2.2  grid_4.5.2         </span></span>
<span><span class="co">##  [79] spdep_1.4-1         impute_1.84.0       tidyr_1.3.2        </span></span>
<span><span class="co">##  [82] ggforce_0.5.0       Formula_1.2-5       cli_3.6.5          </span></span>
<span><span class="co">##  [85] textshaping_1.0.4   S4Arrays_1.10.1     viridisLite_0.4.2  </span></span>
<span><span class="co">##  [88] dplyr_1.1.4         gtable_0.3.6        rstatix_0.7.3      </span></span>
<span><span class="co">##  [91] sass_0.4.10         digest_0.6.39       classInt_0.4-11    </span></span>
<span><span class="co">##  [94] SparseArray_1.10.8  ggrepel_0.9.6       rjson_0.2.23       </span></span>
<span><span class="co">##  [97] htmlwidgets_1.6.4   farver_2.1.2        memoise_2.0.1      </span></span>
<span><span class="co">## [100] htmltools_0.5.9     pkgdown_2.2.0       lifecycle_1.0.5    </span></span>
<span><span class="co">## [103] httr_1.4.7          statmod_1.5.1       MASS_7.3-65</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sara Gosline.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
