<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>spammR microbiome evaluation • spammR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="spammR microbiome evaluation">
<meta name="description" content="A basic walkthrough of how to use the spammR package to perform a
spatial omics analysis for microbiome data.
">
<meta property="og:description" content="A basic walkthrough of how to use the spammR package to perform a
spatial omics analysis for microbiome data.
">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spammR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.17</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/lipidProt.html">spammR lipidomic and proteomic integration</a></li>
    <li><a class="dropdown-item" href="../articles/spatMicrobiome.html">spammR microbiome evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/spatProt.html">spammR Spatial Proteomics Example</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>spammR microbiome evaluation</h1>
                        <h4 data-toc-skip class="author">Yannick
Mahlich</h4>
            
            <h4 data-toc-skip class="date">Feb 04, 2026</h4>
      

      <div class="d-none name"><code>spatMicrobiome.Rmd</code></div>
    </div>

    
    
<p>This vignette walks you through an example of using spammR for
metagenomic data across a geographic area.</p>
<div class="section level2">
<h2 id="r-and-other-dependencies">R and other dependencies<a class="anchor" aria-label="anchor" href="#r-and-other-dependencies"></a>
</h2>
<p>We require a few packages for the mapping requirements that are
beyond the basic spammR tools.</p>
<p>The following R packages are required: - <a href="https://github.com/PNNL-CompBio/spammR" class="external-link">spammR</a> - <a href="https://r-spatial.github.io/sf/index.html" class="external-link">sf</a> relies on other
non R-libraries (see <a href="https://r-spatial.github.io/sf/index.html#installing" class="external-link">Installing</a>
for more details on how to install sf including the dependencies)</p>
<p>The following non-R packages may be required: - <a href="https://libgeos.org/usage/install/" class="external-link">GEOS</a> - <a href="https://proj.org/en/stable/install.html" class="external-link">PROJ</a> - <a href="https://gdal.org/en/stable/download.html" class="external-link">GDAL</a></p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">spammR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/" class="external-link">tibble</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="co">## these two libraries are only used for geographic data</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-spatial.github.io/sf/" class="external-link">sf</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rspatial.org/" class="external-link">terra</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="metagenomic-data-formatting-requirements">Metagenomic data formatting requirements<a class="anchor" aria-label="anchor" href="#metagenomic-data-formatting-requirements"></a>
</h2>
<p>The data that we will be working with is a <a href="https://www.genome.jp/kegg/ko.html" class="external-link">KEGG ortholog</a> enrichment
from the 1000 soils project. The <a href="https://www.emsl.pnnl.gov/project/60141" class="external-link">1000 soils project</a>
has the nice feature that each location has two sampling at different
depths (‘top’ and ‘bottom’). We will use this circumstance in
combination with the KO enrichment to generate a pathway enrichment for
each sample location and depth and then visualize the differential
between the two depths.</p>
<p>We have downloaded the 1000 soil data and stored it as two separate
files that can be loaded directly in the package.</p>
<p>The first file is the actual KO measurements.</p>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>We can then create a metadata file that The KO metadata table will
contain a mapping from KO identifier - row names in the Omics
Measurement table - to a more descriptive annotation.For testing
purposes this currently is only the row names &amp; incremented integer
values.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oneKSoilsOmicsMeta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>KO <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">character</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">oneKSoilsData</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">oneKSoilsOmicsMeta</span><span class="op">[</span><span class="st">"KO"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">oneKSoilsData</span><span class="op">)</span></span>
<span><span class="va">oneKSoilsOmicsMeta</span><span class="op">[</span><span class="st">"annotation"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">1</span>, by <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                        length.out <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">oneKSoilsOmicsMeta</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">oneKSoilsOmicsMeta</span><span class="op">[</span><span class="st">"annotation"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">oneKSoilsOmicsMeta</span><span class="op">[</span><span class="st">"annotation"</span><span class="op">]</span>, </span>
<span>                                           <span class="va">as.character</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">oneKSoilsOmicsMeta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##       KO annotation</span></span>
<span><span class="co">## 1 K00001          1</span></span>
<span><span class="co">## 2 K00002          2</span></span>
<span><span class="co">## 3 K00003          3</span></span>
<span><span class="co">## 4 K00004          4</span></span>
<span><span class="co">## 5 K00005          5</span></span>
<span><span class="co">## 6 K00006          6</span></span></code></pre>
<p>Next we need to import the lat/long information of the locations
where the 1000 soil project samples where extracted. The file was
retrieved from the <a href="https://shinyproxy.emsl.pnnl.gov/app/monet-vis-testbed" class="external-link">1000 Soils
Project Shiny App</a>. The metadata was retrieved by using the
Query&gt;Information tab. On the left side (the “Information Menu”) we
select <code>SAMPLE_ID</code>, <code>latitude</code> &amp;
<code>longitude</code> and download the resulting mapping.</p>
<p>Not all <code>Sample_ID</code> entries have complete lat/long data.
To generate a <code>sf</code> object (see below) which will be needed as
input to create a <code>terra:SpatRaster</code> object the input table
can only contain complete cases (i.e. no <code>NaN</code> values). The
sample metadata contains information about the type of Biome, the layer
of the soil sample (‘top’ or ‘bottom’), and the pH, for example, as well
as the coordinates.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"oneKSoilsMeta"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">oneKSoilsMeta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Sample_ID Site_Code Core_Layer latitude  longitude Elev.m Soil.Temperature.C</span></span>
<span><span class="co">## 1  ANZA_TOP      ANZA        TOP 33.30533 -116.25401    191               42.7</span></span>
<span><span class="co">## 2  ANZA_BTM      ANZA        BTM 33.30533 -116.25401    191               42.7</span></span>
<span><span class="co">## 3  BLAN_TOP      BLAN        TOP 39.05941  -78.07207    174               19.7</span></span>
<span><span class="co">## 4  BLAN_BTM      BLAN        BTM 39.05941  -78.07207    174               19.7</span></span>
<span><span class="co">## 5  CFS1_TOP      CFS1        TOP 35.38280  -78.03818     24               18.5</span></span>
<span><span class="co">## 6  CFS1_BTM      CFS1        BTM 35.38280  -78.03818     24               18.5</span></span>
<span><span class="co">##   VolWaterContent.m3.m3   pH K_mg_per_kg SO4.S_mg_per_kg B_mg_per_kg</span></span>
<span><span class="co">## 1                0.0490 8.75         107               4        0.25</span></span>
<span><span class="co">## 2                0.0490 9.20          82               4        0.20</span></span>
<span><span class="co">## 3                0.1740 6.50         189              10        0.42</span></span>
<span><span class="co">## 4                0.1740 5.92         122              19        0.27</span></span>
<span><span class="co">## 5                0.0515 6.66         231              NA          NA</span></span>
<span><span class="co">## 6                0.0515 6.98         107               6        0.26</span></span>
<span><span class="co">##   Zn_mg_per_kg Mn_mg_per_kg Cu_mg_per_kg Fe_mg_per_kg Ca_Meq_per_100_g</span></span>
<span><span class="co">## 1          0.2          1.9          0.1           13             3.27</span></span>
<span><span class="co">## 2          0.3          1.1          0.1           14             3.05</span></span>
<span><span class="co">## 3          0.7         28.3          0.7           54             5.47</span></span>
<span><span class="co">## 4          0.2         15.9          0.5           16             4.31</span></span>
<span><span class="co">## 5           NA           NA           NA           NA             2.41</span></span>
<span><span class="co">## 6          0.3          0.7          0.5           18             2.93</span></span>
<span><span class="co">##   Mg_Meq_per_100_g Na_Meq_per_100_g BIOME                          biome_name</span></span>
<span><span class="co">## 1             0.26             0.10    13          Deserts &amp; Xeric Shrublands</span></span>
<span><span class="co">## 2             0.42             0.08    13          Deserts &amp; Xeric Shrublands</span></span>
<span><span class="co">## 3             0.81             0.05     4 Temperate Broadleaf &amp; Mixed Forests</span></span>
<span><span class="co">## 4             0.57             0.06     4 Temperate Broadleaf &amp; Mixed Forests</span></span>
<span><span class="co">## 5             0.68             0.08     5           Temperate Conifer Forests</span></span>
<span><span class="co">## 6             1.14             0.06     5           Temperate Conifer Forests</span></span></code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot2</span><span class="fu">::</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">oneKSoilsMeta</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Core_Layer</span>, y <span class="op">=</span> <span class="va">pH</span>, fill <span class="op">=</span> <span class="va">biome_name</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 4 rows containing non-finite outside the scale range</span></span>
<span><span class="co">## (`stat_boxplot()`).</span></span></code></pre>
<p><img src="spatMicrobiome_files/figure-html/coordiantes-1.png" class="r-plt" alt="" width="700"></p>
</div>
<div class="section level2">
<h2 id="geographic-map-formatting-requirements">Geographic map formatting requirements<a class="anchor" aria-label="anchor" href="#geographic-map-formatting-requirements"></a>
</h2>
<p>Maps can be downloaded from public repositories as a ‘shape’ file
that can then be processed for coordinates and visualization. Here we
show how to retrieve a shape file d from the <a href="https://www.census.gov/cgi-bin/geo/shapefiles/index.php" class="external-link">US Census
Bureau</a> with the following parameters:</p>
<ul>
<li>Year = 2024</li>
<li>Layer type = “States (and equivalent)”</li>
</ul>
<p>Or downloaded in a zip file from the [census FTP site] (<a href="https://www2.census.gov/geo/tiger/TIGER2024/STATE/" class="external-link uri">https://www2.census.gov/geo/tiger/TIGER2024/STATE/</a>).</p>
<p>The code chunk below will not be executed but is displayed as example
of how to load a downloaded shape file using the sf package.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">us_map</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_read.html" class="external-link">st_read</a></span><span class="op">(</span><span class="st">"../data/tl_2024_us_state/tl_2024_us_state.shp"</span>, </span>
<span>                      quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">us_map</span></span></code></pre></div>
<p>Instead will load the file from the supplied <code>rda</code>
file:</p>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>This map contains the entire US, so for visualization purposes we
focus only on the continental US.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">us_continental</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"WV"</span>, <span class="st">"FL"</span>, <span class="st">"IL"</span>, <span class="st">"MN"</span>, <span class="st">"MD"</span>, <span class="st">"RI"</span>, <span class="st">"ID"</span>, <span class="st">"NH"</span>, <span class="st">"NC"</span>, <span class="st">"VT"</span>,</span>
<span>  <span class="st">"CT"</span>, <span class="st">"DE"</span>, <span class="st">"NM"</span>, <span class="st">"CA"</span>, <span class="st">"NJ"</span>, <span class="st">"WI"</span>, <span class="st">"OR"</span>, <span class="st">"NE"</span>, <span class="st">"PA"</span>, <span class="st">"WA"</span>,</span>
<span>  <span class="st">"LA"</span>, <span class="st">"GA"</span>, <span class="st">"AL"</span>, <span class="st">"UT"</span>, <span class="st">"OH"</span>, <span class="st">"TX"</span>, <span class="st">"CO"</span>, <span class="st">"SC"</span>, <span class="st">"OK"</span>, <span class="st">"TN"</span>,</span>
<span>  <span class="st">"WY"</span>, <span class="st">"ND"</span>, <span class="st">"KY"</span>, <span class="st">"ME"</span>, <span class="st">"NY"</span>, <span class="st">"NV"</span>, <span class="st">"MI"</span>, <span class="st">"AR"</span>, <span class="st">"MS"</span>, <span class="st">"MO"</span>,</span>
<span>  <span class="st">"MT"</span>, <span class="st">"KS"</span>, <span class="st">"IN"</span>, <span class="st">"SD"</span>, <span class="st">"MA"</span>, <span class="st">"VA"</span>, <span class="st">"DC"</span>, <span class="st">"IA"</span>, <span class="st">"AZ"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">s_us_cont</span> <span class="op">&lt;-</span> <span class="va">us_map</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="va">.</span><span class="op">[</span><span class="va">.</span><span class="op">$</span><span class="va">STUSPS</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="va">us_continental</span>, <span class="op">]</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/vect.html" class="external-link">vect</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">s_us_cont</span></span></code></pre></div>
<pre><code><span><span class="co">##  class       : SpatVector </span></span>
<span><span class="co">##  geometry    : polygons </span></span>
<span><span class="co">##  dimensions  : 49, 15  (geometries, attributes)</span></span>
<span><span class="co">##  extent      : -124.849, -66.88544, 24.39631, 49.38448  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">##  coord. ref. : lon/lat NAD83 (EPSG:4269) </span></span>
<span><span class="co">##  names       : REGION DIVISION STATEFP  STATENS GEOID     GEOIDFQ STUSPS</span></span>
<span><span class="co">##  type        :  &lt;chr&gt;    &lt;chr&gt;   &lt;chr&gt;    &lt;chr&gt; &lt;chr&gt;       &lt;chr&gt;  &lt;chr&gt;</span></span>
<span><span class="co">##  values      :      3        5      54 01779805    54 0400000US54     WV</span></span>
<span><span class="co">##                     3        5      12 00294478    12 0400000US12     FL</span></span>
<span><span class="co">##                     2        3      17 01779784    17 0400000US17     IL</span></span>
<span><span class="co">##           NAME  LSAD MTFCC (and 5 more)</span></span>
<span><span class="co">##          &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;             </span></span>
<span><span class="co">##  West Virginia    00 G4000             </span></span>
<span><span class="co">##        Florida    00 G4000             </span></span>
<span><span class="co">##       Illinois    00 G4000</span></span></code></pre>
<p>Now that we have the coordinate information we need to convert it to
pixels using the <code>terra</code> package, which requires creating a
rasterization template using the” <code><a href="https://rspatial.github.io/terra/reference/SpatRaster-class.html" class="external-link">terra::SpatRaster</a></code>. From
this object we can generate a rasterized image while retaining the
location information and ability to project the map into a different map
projection.</p>
<p>The <code>terra</code> package can read in the shape file as
follows:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">template</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/rast.html" class="external-link">rast</a></span><span class="op">(</span></span>
<span>  <span class="va">s_us_cont</span>,</span>
<span>  res <span class="op">=</span> <span class="fl">0.01</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>And then create a rasterized landmass of the continental US:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># us_raster &lt;- rasterize(s_us_cont, template, background=0)</span></span>
<span><span class="co"># us_raster &lt;- rasterize(as.lines(s_us_cont), template, field='STATEFP', </span></span>
<span><span class="co"># touches=TRUE)</span></span>
<span></span>
<span><span class="va">us_raster</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/rasterize.html" class="external-link">rasterize</a></span><span class="op">(</span><span class="va">s_us_cont</span>, <span class="va">template</span>, field <span class="op">=</span> <span class="st">"STATEFP"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="mapping-soil-locations-to-map">Mapping soil locations to map<a class="anchor" aria-label="anchor" href="#mapping-soil-locations-to-map"></a>
</h3>
<p>Next we generate an <code>sf</code> object from the soils coordinate
data so that we can then map it to the same rasterized image as the US
map object:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">oneKSoilsMeta</span> <span class="op">&lt;-</span> <span class="va">oneKSoilsMeta</span><span class="op">[</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">oneKSoilsMeta</span><span class="op">$</span><span class="va">latitude</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>                                 <span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">oneKSoilsMeta</span><span class="op">$</span><span class="va">longitude</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="va">coords</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sf.html" class="external-link">st_as_sf</a></span><span class="op">(</span><span class="va">oneKSoilsMeta</span>, coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"longitude"</span>, <span class="st">"latitude"</span><span class="op">)</span>, </span>
<span>                       remove <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_crs.html" class="external-link">st_crs</a></span><span class="op">(</span><span class="va">coords</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_crs.html" class="external-link">st_crs</a></span><span class="op">(</span><span class="va">s_us_cont</span><span class="op">)</span></span>
<span><span class="va">coords</span></span></code></pre></div>
<pre><code><span><span class="co">## Simple feature collection with 127 features and 19 fields</span></span>
<span><span class="co">## Geometry type: POINT</span></span>
<span><span class="co">## Dimension:     XY</span></span>
<span><span class="co">## Bounding box:  xmin: -121.9607 ymin: 28.12584 xmax: -70.63536 ymax: 47.16306</span></span>
<span><span class="co">## Geodetic CRS:  NAD83</span></span>
<span><span class="co">## First 10 features:</span></span>
<span><span class="co">##    Sample_ID Site_Code Core_Layer Elev.m Soil.Temperature.C</span></span>
<span><span class="co">## 1   ANZA_TOP      ANZA        TOP    191               42.7</span></span>
<span><span class="co">## 2   ANZA_BTM      ANZA        BTM    191               42.7</span></span>
<span><span class="co">## 3   BLAN_TOP      BLAN        TOP    174               19.7</span></span>
<span><span class="co">## 4   BLAN_BTM      BLAN        BTM    174               19.7</span></span>
<span><span class="co">## 5   CFS1_TOP      CFS1        TOP     24               18.5</span></span>
<span><span class="co">## 6   CFS1_BTM      CFS1        BTM     24               18.5</span></span>
<span><span class="co">## 7   CFS2_TOP      CFS2        TOP     24               23.0</span></span>
<span><span class="co">## 8   CFS2_BTM      CFS2        BTM     24               23.0</span></span>
<span><span class="co">## 9   CLBJ_BTM      CLBJ        BTM    274               18.4</span></span>
<span><span class="co">## 10  CLBJ_TOP      CLBJ        TOP    274               18.4</span></span>
<span><span class="co">##    VolWaterContent.m3.m3   pH K_mg_per_kg SO4.S_mg_per_kg B_mg_per_kg</span></span>
<span><span class="co">## 1                 0.0490 8.75         107               4        0.25</span></span>
<span><span class="co">## 2                 0.0490 9.20          82               4        0.20</span></span>
<span><span class="co">## 3                 0.1740 6.50         189              10        0.42</span></span>
<span><span class="co">## 4                 0.1740 5.92         122              19        0.27</span></span>
<span><span class="co">## 5                 0.0515 6.66         231              NA          NA</span></span>
<span><span class="co">## 6                 0.0515 6.98         107               6        0.26</span></span>
<span><span class="co">## 7                 0.0695 6.66         134               4        0.11</span></span>
<span><span class="co">## 8                 0.0695 6.78          66               2        0.06</span></span>
<span><span class="co">## 9                 0.3380 4.94         144               5        0.18</span></span>
<span><span class="co">## 10                0.3380 7.28         156               6        0.40</span></span>
<span><span class="co">##    Zn_mg_per_kg Mn_mg_per_kg Cu_mg_per_kg Fe_mg_per_kg Ca_Meq_per_100_g</span></span>
<span><span class="co">## 1           0.2          1.9          0.1           13             3.27</span></span>
<span><span class="co">## 2           0.3          1.1          0.1           14             3.05</span></span>
<span><span class="co">## 3           0.7         28.3          0.7           54             5.47</span></span>
<span><span class="co">## 4           0.2         15.9          0.5           16             4.31</span></span>
<span><span class="co">## 5            NA           NA           NA           NA             2.41</span></span>
<span><span class="co">## 6           0.3          0.7          0.5           18             2.93</span></span>
<span><span class="co">## 7           3.6          5.1          2.2           48             1.76</span></span>
<span><span class="co">## 8           0.3          2.0          0.3           17             1.13</span></span>
<span><span class="co">## 9           0.4          9.6          0.2           21             1.42</span></span>
<span><span class="co">## 10          1.6          5.9          0.2           12             9.02</span></span>
<span><span class="co">##    Mg_Meq_per_100_g Na_Meq_per_100_g BIOME</span></span>
<span><span class="co">## 1              0.26             0.10    13</span></span>
<span><span class="co">## 2              0.42             0.08    13</span></span>
<span><span class="co">## 3              0.81             0.05     4</span></span>
<span><span class="co">## 4              0.57             0.06     4</span></span>
<span><span class="co">## 5              0.68             0.08     5</span></span>
<span><span class="co">## 6              1.14             0.06     5</span></span>
<span><span class="co">## 7              0.50             0.05     4</span></span>
<span><span class="co">## 8              0.32             0.04     4</span></span>
<span><span class="co">## 9              1.08             0.10     8</span></span>
<span><span class="co">## 10             1.04             0.07     8</span></span>
<span><span class="co">##                                     biome_name                   geometry</span></span>
<span><span class="co">## 1                   Deserts &amp; Xeric Shrublands  POINT (-116.254 33.30533)</span></span>
<span><span class="co">## 2                   Deserts &amp; Xeric Shrublands  POINT (-116.254 33.30533)</span></span>
<span><span class="co">## 3          Temperate Broadleaf &amp; Mixed Forests POINT (-78.07207 39.05941)</span></span>
<span><span class="co">## 4          Temperate Broadleaf &amp; Mixed Forests POINT (-78.07207 39.05941)</span></span>
<span><span class="co">## 5                    Temperate Conifer Forests  POINT (-78.03818 35.3828)</span></span>
<span><span class="co">## 6                    Temperate Conifer Forests  POINT (-78.03818 35.3828)</span></span>
<span><span class="co">## 7          Temperate Broadleaf &amp; Mixed Forests POINT (-81.43417 35.38329)</span></span>
<span><span class="co">## 8          Temperate Broadleaf &amp; Mixed Forests POINT (-81.43417 35.38329)</span></span>
<span><span class="co">## 9  Temperate Grasslands, Savannas &amp; Shrublands POINT (-97.57043 33.40164)</span></span>
<span><span class="co">## 10 Temperate Grasslands, Savannas &amp; Shrublands POINT (-97.57043 33.40164)</span></span></code></pre>
<p>Using the same rasterization template as we used above, we can make
the soils coordinates fit on pixes on the US map.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coords_rast</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/rasterize.html" class="external-link">rasterize</a></span><span class="op">(</span><span class="va">coords</span>, <span class="va">template</span><span class="op">)</span></span>
<span><span class="va">coords_rast</span></span></code></pre></div>
<pre><code><span><span class="co">## class       : SpatRaster </span></span>
<span><span class="co">## size        : 2499, 5796, 1  (nrow, ncol, nlyr)</span></span>
<span><span class="co">## resolution  : 0.01, 0.01  (x, y)</span></span>
<span><span class="co">## extent      : -124.849, -66.88897, 24.39631, 49.38631  (xmin, xmax, ymin, ymax)</span></span>
<span><span class="co">## coord. ref. : lon/lat NAD83 (EPSG:4269) </span></span>
<span><span class="co">## source(s)   : memory</span></span>
<span><span class="co">## name        : last </span></span>
<span><span class="co">## min value   :    1 </span></span>
<span><span class="co">## max value   :    1</span></span></code></pre>
<div class="section level4">
<h4 id="get-the-pixel-xy-not-latlong-xy-of-metagenome-samples">get the “Pixel” x/y (NOT lat/long x/y) of metagenome samples<a class="anchor" aria-label="anchor" href="#get-the-pixel-xy-not-latlong-xy-of-metagenome-samples"></a>
</h4>
<p>Finally to retrieve the “pixel” coordinates instead of the x/y
(lat/long) coordinates that are recorded in the
<code><a href="https://rspatial.github.io/terra/reference/SpatRaster-class.html" class="external-link">terra::SpatRaster</a></code> objects the following procedure is
employed:</p>
<ul>
<li>use <code><a href="https://magrittr.tidyverse.org/reference/aliases.html" class="external-link">extract()</a></code> to get the cells which the coordinates
are projected into in the raster image</li>
<li>use <code><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply()</a></code> get the raster coordinates (‘pixel’ x/y)
for each row with <code><a href="https://rspatial.github.io/terra/reference/xyCellFrom.html" class="external-link">terra::rowColFromCell()</a></code>
</li>
<li>use <code><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind()</a></code> to combine the cells, ‘pixel’ x/y and
original sample coordinate <code>data.frame</code>
</li>
<li>use <code><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">dplyr::rename()</a></code> to clean up
<code>data.frame</code>
</li>
</ul>
<p>This process does come with slight differences in lat/long. Most
likely due to the rasterization process?</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">coords_rast</span>, <span class="va">coords</span>, xy <span class="op">=</span> <span class="cn">TRUE</span>, cells <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cells</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   ID last    cell          x        y</span></span>
<span><span class="co">## 1  1    1 9320828 -116.25397 33.30131</span></span>
<span><span class="co">## 2  2    1 9320828 -116.25397 33.30131</span></span>
<span><span class="co">## 3  3    1 5986150  -78.07397 39.06131</span></span>
<span><span class="co">## 4  4    1 5986150  -78.07397 39.06131</span></span>
<span><span class="co">## 5  5    1 8119082  -78.03397 35.38131</span></span>
<span><span class="co">## 6  6    1 8119082  -78.03397 35.38131</span></span></code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cells_tmp</span> <span class="op">&lt;-</span> <span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">coords_rast</span>, <span class="va">coords</span>, xy <span class="op">=</span> <span class="cn">TRUE</span>, cells <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">xy_rast</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">cells</span><span class="op">[</span><span class="st">"cell"</span><span class="op">]</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/xyCellFrom.html" class="external-link">rowColFromCell</a></span><span class="op">(</span><span class="va">coords_rast</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cells</span> <span class="op">&lt;-</span> <span class="va">cells_tmp</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">oneKSoilsMeta</span>, <span class="va">xy_rast</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">.</span>, y_pixels <span class="op">=</span> <span class="st">"cell.1"</span>, x_pixels <span class="op">=</span> <span class="st">"cell.2"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">.</span>, select <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">y</span>, <span class="va">ID</span>, <span class="va">last</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">cells</span><span class="op">$</span><span class="va">y_pixels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">coords_rast</span><span class="op">)</span> <span class="op">-</span> <span class="va">cells</span><span class="op">$</span><span class="va">y_pixels</span></span></code></pre></div>
<p>Those need to be finally written out / transferred into the
<code>spammR</code> package for further use.</p>
</div>
<div class="section level4">
<h4 id="exporting-the-rasterized-map">Exporting the rasterized map<a class="anchor" aria-label="anchor" href="#exporting-the-rasterized-map"></a>
</h4>
<p>The rasterized map can be exported with the command below.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">terra</span><span class="fu">::</span><span class="fu"><a href="https://rspatial.github.io/terra/reference/writeRaster.html" class="external-link">writeRaster</a></span><span class="op">(</span><span class="va">us_raster</span>, <span class="st">"us_map.png"</span>, NAflag <span class="op">=</span> <span class="fl">0</span>, overwrite <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="examplary-plotting">Examplary plotting<a class="anchor" aria-label="anchor" href="#examplary-plotting"></a>
</h4>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coords_test</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_as_sf.html" class="external-link">st_as_sf</a></span><span class="op">(</span><span class="va">oneKSoilsMeta</span>, </span>
<span>                            coords <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"longitude"</span>, <span class="st">"latitude"</span><span class="op">)</span>, </span>
<span>                            remove <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_crs.html" class="external-link">st_crs</a></span><span class="op">(</span><span class="va">coords_test</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">sf</span><span class="fu">::</span><span class="fu"><a href="https://r-spatial.github.io/sf/reference/st_crs.html" class="external-link">st_crs</a></span><span class="op">(</span><span class="va">s_us_cont</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">us_raster</span>, col <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span><span class="co"># lines(as.polygons(us_raster), col='black')</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/plot.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">coords_test</span>, col <span class="op">=</span> <span class="st">"darkred"</span>, add <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning in plot.sf(coords_test, col = "darkred", add = TRUE): ignoring all but</span></span>
<span><span class="co">## the first attribute</span></span></code></pre>
<p><img src="spatMicrobiome_files/figure-html/unnamed-chunk-6-1.png" class="r-plt" alt="" width="700"></p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="using-spammr-on-metagenomic-geographic-data">Using <code>spammR</code> on metagenomic geographic data<a class="anchor" aria-label="anchor" href="#using-spammr-on-metagenomic-geographic-data"></a>
</h2>
<p>Now that the data are properly formatted we can put them into a
<code>SpatialExperiment</code> object that is needed by spammR. The data
is comprised of:</p>
<ul>
<li>Omics Measurements: species abundances across regions</li>
<li>Omics Metadata: metadata from these species</li>
<li>Sample Metadata: information about the sample location</li>
<li>Image files (generated by the above code): file representing image
sample</li>
</ul>
<div class="section level3">
<h3 id="sample-metadata">Sample Metadata<a class="anchor" aria-label="anchor" href="#sample-metadata"></a>
</h3>
<p>Here we build the sample metadata table given the previously
generated coordinate raster during the rasterization of the map. The
SpatialExperiment object requires information about the x and y
coordinates and information about each spot in the image for which we
have omics data.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dm</span> <span class="op">&lt;-</span> <span class="va">cells</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">dm</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">dm</span><span class="op">$</span><span class="va">Sample_ID</span></span>
<span><span class="co"># dm[,1] &lt;- NULL</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"cell"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"longitude.1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"latitude.1"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"Image"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"x_max"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">coords_rast</span><span class="op">)</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"y_max"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">coords_rast</span><span class="op">)</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"x_origin"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"y_origin"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"spot_height"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"spot_width"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">500</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"above_40_deg_lat"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">dm</span><span class="op">[</span>, <span class="st">"latitude"</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">40</span>, </span>
<span>                                                  true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"psychrophile"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">dm</span><span class="op">[</span>, <span class="st">"Soil.Temperature.C"</span><span class="op">]</span> <span class="op">&lt;=</span> <span class="fl">15</span>, </span>
<span>                                       true <span class="op">=</span> <span class="fl">1</span>, false <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"Core_Layer_bin"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">dm</span><span class="op">[</span>, <span class="st">"Core_Layer"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"TOP"</span>, </span>
<span>                                         true <span class="op">=</span> <span class="st">"TOP"</span>, false <span class="op">=</span> <span class="st">"BTM"</span><span class="op">)</span></span>
<span><span class="va">dm</span><span class="op">[</span>, <span class="st">"Desert"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/if_else.html" class="external-link">if_else</a></span><span class="op">(</span><span class="va">dm</span><span class="op">[</span>, <span class="st">"BIOME"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"13"</span>, </span>
<span>                                 true <span class="op">=</span> <span class="st">"Desert"</span>, false <span class="op">=</span> <span class="st">"NonDesert"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="va">dm</span></span>
<span><span class="fu"><a href="https://rspatial.github.io/terra/reference/headtail.html" class="external-link">head</a></span><span class="op">(</span><span class="va">data_meta</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          Sample_ID Site_Code Core_Layer latitude  longitude Elev.m</span></span>
<span><span class="co">## ANZA_TOP  ANZA_TOP      ANZA        TOP 33.30533 -116.25401    191</span></span>
<span><span class="co">## ANZA_BTM  ANZA_BTM      ANZA        BTM 33.30533 -116.25401    191</span></span>
<span><span class="co">## BLAN_TOP  BLAN_TOP      BLAN        TOP 39.05941  -78.07207    174</span></span>
<span><span class="co">## BLAN_BTM  BLAN_BTM      BLAN        BTM 39.05941  -78.07207    174</span></span>
<span><span class="co">## CFS1_TOP  CFS1_TOP      CFS1        TOP 35.38280  -78.03818     24</span></span>
<span><span class="co">## CFS1_BTM  CFS1_BTM      CFS1        BTM 35.38280  -78.03818     24</span></span>
<span><span class="co">##          Soil.Temperature.C VolWaterContent.m3.m3   pH K_mg_per_kg</span></span>
<span><span class="co">## ANZA_TOP               42.7                0.0490 8.75         107</span></span>
<span><span class="co">## ANZA_BTM               42.7                0.0490 9.20          82</span></span>
<span><span class="co">## BLAN_TOP               19.7                0.1740 6.50         189</span></span>
<span><span class="co">## BLAN_BTM               19.7                0.1740 5.92         122</span></span>
<span><span class="co">## CFS1_TOP               18.5                0.0515 6.66         231</span></span>
<span><span class="co">## CFS1_BTM               18.5                0.0515 6.98         107</span></span>
<span><span class="co">##          SO4.S_mg_per_kg B_mg_per_kg Zn_mg_per_kg Mn_mg_per_kg Cu_mg_per_kg</span></span>
<span><span class="co">## ANZA_TOP               4        0.25          0.2          1.9          0.1</span></span>
<span><span class="co">## ANZA_BTM               4        0.20          0.3          1.1          0.1</span></span>
<span><span class="co">## BLAN_TOP              10        0.42          0.7         28.3          0.7</span></span>
<span><span class="co">## BLAN_BTM              19        0.27          0.2         15.9          0.5</span></span>
<span><span class="co">## CFS1_TOP              NA          NA           NA           NA           NA</span></span>
<span><span class="co">## CFS1_BTM               6        0.26          0.3          0.7          0.5</span></span>
<span><span class="co">##          Fe_mg_per_kg Ca_Meq_per_100_g Mg_Meq_per_100_g Na_Meq_per_100_g BIOME</span></span>
<span><span class="co">## ANZA_TOP           13             3.27             0.26             0.10    13</span></span>
<span><span class="co">## ANZA_BTM           14             3.05             0.42             0.08    13</span></span>
<span><span class="co">## BLAN_TOP           54             5.47             0.81             0.05     4</span></span>
<span><span class="co">## BLAN_BTM           16             4.31             0.57             0.06     4</span></span>
<span><span class="co">## CFS1_TOP           NA             2.41             0.68             0.08     5</span></span>
<span><span class="co">## CFS1_BTM           18             2.93             1.14             0.06     5</span></span>
<span><span class="co">##                                   biome_name y_pixels x_pixels Image x_max</span></span>
<span><span class="co">## ANZA_TOP          Deserts &amp; Xeric Shrublands      890      860     0  5796</span></span>
<span><span class="co">## ANZA_BTM          Deserts &amp; Xeric Shrublands      890      860     0  5796</span></span>
<span><span class="co">## BLAN_TOP Temperate Broadleaf &amp; Mixed Forests     1466     4678     0  5796</span></span>
<span><span class="co">## BLAN_BTM Temperate Broadleaf &amp; Mixed Forests     1466     4678     0  5796</span></span>
<span><span class="co">## CFS1_TOP           Temperate Conifer Forests     1098     4682     0  5796</span></span>
<span><span class="co">## CFS1_BTM           Temperate Conifer Forests     1098     4682     0  5796</span></span>
<span><span class="co">##          y_max x_origin y_origin spot_height spot_width above_40_deg_lat</span></span>
<span><span class="co">## ANZA_TOP  2499        0        0         500        500                0</span></span>
<span><span class="co">## ANZA_BTM  2499        0        0         500        500                0</span></span>
<span><span class="co">## BLAN_TOP  2499        0        0         500        500                0</span></span>
<span><span class="co">## BLAN_BTM  2499        0        0         500        500                0</span></span>
<span><span class="co">## CFS1_TOP  2499        0        0         500        500                0</span></span>
<span><span class="co">## CFS1_BTM  2499        0        0         500        500                0</span></span>
<span><span class="co">##          psychrophile Core_Layer_bin    Desert</span></span>
<span><span class="co">## ANZA_TOP            0            TOP    Desert</span></span>
<span><span class="co">## ANZA_BTM            0            BTM    Desert</span></span>
<span><span class="co">## BLAN_TOP            0            TOP NonDesert</span></span>
<span><span class="co">## BLAN_BTM            0            BTM NonDesert</span></span>
<span><span class="co">## CFS1_TOP            0            TOP NonDesert</span></span>
<span><span class="co">## CFS1_BTM            0            BTM NonDesert</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="cleaning-up-of-data">CLEANING UP OF DATA<a class="anchor" aria-label="anchor" href="#cleaning-up-of-data"></a>
</h3>
<p>Limiting our selfs to only the top layer of the soil column.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">data_meta</span>, <span class="va">Sample_ID</span> <span class="op"><a href="https://rspatial.github.io/terra/reference/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">oneKSoilsData</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">oneKSoilsData</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">oneKSoilsData</span>, select <span class="op">=</span> <span class="va">data_meta</span><span class="op">$</span><span class="va">Sample_ID</span><span class="op">)</span></span>
<span><span class="va">data_meta</span> <span class="op">&lt;-</span> <span class="va">data_meta</span><span class="op">[</span><span class="va">data_meta</span><span class="op">$</span><span class="va">Core_Layer</span> <span class="op">==</span> <span class="st">"TOP"</span>, <span class="op">]</span></span>
<span><span class="va">oneKSoilsData</span> <span class="op">&lt;-</span> <span class="va">oneKSoilsData</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"TOP"</span>, x <span class="op">=</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">oneKSoilsData</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="create-spatialexperiment-object">Create <code>SpatialExperiment</code> object<a class="anchor" aria-label="anchor" href="#create-spatialexperiment-object"></a>
</h3>
<p>Now that our metadata is formatted correctly we can create the
spatialExperiment object.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">microbiome.spe</span> <span class="op">&lt;-</span> <span class="fu">spammR</span><span class="fu">::</span><span class="fu"><a href="../reference/convert_to_spe.html">convert_to_spe</a></span><span class="op">(</span></span>
<span>  <span class="va">oneKSoilsData</span>, <span class="co">## pooled data table</span></span>
<span>  <span class="va">data_meta</span>, <span class="co">## pooled metadata</span></span>
<span>  <span class="va">oneKSoilsOmicsMeta</span>, <span class="co">## protein identifiers</span></span>
<span>  feature_meta_colname <span class="op">=</span> <span class="st">"KO"</span>, <span class="co"># column name</span></span>
<span>  spatial_coords_colnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"x_pixels"</span>, <span class="st">"y_pixels"</span><span class="op">)</span>,</span>
<span>  image_files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"us_map.png"</span><span class="op">)</span>,</span>
<span>  image_sample_ids <span class="op">=</span> <span class="st">"1000 Soils"</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"1000 Soils"</span>,</span>
<span>  image_ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0"</span><span class="op">)</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"KO"</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="metagenomic-differential-expression-and-enrichment">Metagenomic differential expression and enrichment<a class="anchor" aria-label="anchor" href="#metagenomic-differential-expression-and-enrichment"></a>
</h3>
<p>The pathway enrichment analysis is done based on the presence and
absence of KO associations found within the individual metagenomes. As
such we are not strictly working with “gene abundances” that would be
expected for a more traditional differential expression and pathway
enrichment analysis. Conversely the “abundance” measurements of the
individual KOs also distribute differently to what would be expected. To
this end we are relying on the count based differential expression
analysis that is based on <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">limma::voom</a></code>.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">diff_ex</span> <span class="op">&lt;-</span> <span class="fu">spammR</span><span class="fu">::</span><span class="fu"><a href="../reference/calc_spatial_diff_ex.html">calc_spatial_diff_ex</a></span><span class="op">(</span></span>
<span>  <span class="va">microbiome.spe</span>,</span>
<span>  assay_name <span class="op">=</span> <span class="st">"KO"</span>,</span>
<span>  count_based <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  log_transformed <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  category_col <span class="op">=</span> <span class="st">"Desert"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## We found 562 features with a logFC greater than 1 and </span></span>
<span><span class="co">##                  an ajusted p-value less than 0.05</span></span></code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig_ko</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">diff_ex</span><span class="op">)</span>, <span class="va">Desert_vs_NonDesert.adj.P.Val.limma</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span><span class="va">ups</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sig_ko</span>, <span class="va">Desert_vs_NonDesert.logFC.limma</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="va">downs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">sig_ko</span>, <span class="va">Desert_vs_NonDesert.logFC.limma</span> <span class="op">&lt;</span> <span class="fl">0</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste</a></span><span class="op">(</span></span>
<span>  <span class="st">"We found"</span>, <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">sig_ko</span><span class="op">)</span>, <span class="st">"significantly differentally abundant KOs"</span>,</span>
<span>  <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">ups</span><span class="op">)</span>, <span class="st">"upregulated KOs and"</span>, <span class="fu"><a href="https://rspatial.github.io/terra/reference/dimensions.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">downs</span><span class="op">)</span>, <span class="st">"downregulated"</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "We found 575 significantly differentally abundant KOs 483 upregulated KOs and 92 downregulated"</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ups</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">.</span>, <span class="op">-</span><span class="va">Desert_vs_NonDesert.logFC.limma</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 483 × 8</span></span></span>
<span><span class="co">##    KO     annotation Desert_vs_NonDesert.logFC.limma Desert_vs_NonDesert.AveEx…¹</span></span>
<span><span class="co">##    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                                <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>                       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 1</span> K21136 11079                                 4.30                        5.91</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 2</span> K15547 8600                                  4.15                        4.98</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 3</span> K09758 5602                                  4.07                        5.42</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 4</span> K01535 1144                                  4.04                        5.02</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 5</span> K15549 8602                                  4.04                        5.18</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 6</span> K07085 4377                                  3.87                        7.23</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 7</span> K12137 6892                                  3.87                        6.94</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 8</span> K13926 7769                                  3.71                        5.59</span></span>
<span><span class="co">## <span style="color: #BCBCBC;"> 9</span> K05817 3685                                  3.64                        5.74</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">10</span> K21134 11077                                 3.56                        7.72</span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 473 more rows</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ abbreviated name: ¹​Desert_vs_NonDesert.AveExpr.limma</span></span></span>
<span><span class="co">## <span style="color: #949494;"># ℹ 4 more variables: Desert_vs_NonDesert.t.limma &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Desert_vs_NonDesert.P.Value.limma &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Desert_vs_NonDesert.adj.P.Val.limma &lt;dbl&gt;,</span></span></span>
<span><span class="co">## <span style="color: #949494;">#   Desert_vs_NonDesert.B.limma &lt;dbl&gt;</span></span></span></code></pre>
</div>
<div class="section level3">
<h3 id="spatial-heatmap-showing-abundance-of-significant-ko">Spatial heatmap showing abundance of significant KO<a class="anchor" aria-label="anchor" href="#spatial-heatmap-showing-abundance-of-significant-ko"></a>
</h3>
<p>K09758 (aspartate 4-decarboxylase) being one of the “upregulated KOs”
we can visualize the abundances across the different sampling locations
with <code>spammr::spatial_heatmap()</code></p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span></span>
<span>  <span class="va">microbiome.spe</span>,</span>
<span>  feature <span class="op">=</span> <span class="st">"K09758"</span>,</span>
<span>  <span class="co"># feature_type = 'KO',</span></span>
<span>  assay_name <span class="op">=</span> <span class="st">"KO"</span>,</span>
<span>  sample_id <span class="op">=</span> <span class="st">"1000 Soils"</span>,</span>
<span>  image_id <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>  sample_label_size <span class="op">=</span> <span class="fl">2.0</span>, </span>
<span>  title_size <span class="op">=</span> <span class="fl">10</span>, </span>
<span>  label_column <span class="op">=</span> <span class="st">"Desert"</span>,</span>
<span>  metric_display <span class="op">=</span> <span class="st">"KO association abundance"</span>,</span>
<span>  plot_title <span class="op">=</span> <span class="st">"Abundance of aspartate 4-decarboxylase (K09758)"</span>,</span>
<span>  interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 2 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_label()`).</span></span></code></pre>
<p><img src="spatMicrobiome_files/figure-html/heatmap-1.png" class="r-plt" alt="" width="700"></p>
</div>
<div class="section level3">
<h3 id="pathway-enrichment">Pathway enrichment<a class="anchor" aria-label="anchor" href="#pathway-enrichment"></a>
</h3>
<p>For the pathway enrichment part of the anaysis we first need to load
“genesets” in the required format for leapR that will be called
internally. The geneset is effectively a list of KOs that are associated
with individual KEGG pathways.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">kk</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/download.file.html" class="external-link">download.file</a></span><span class="op">(</span><span class="st">"https://api.figshare.com/v2/file/download/55158818"</span>, </span>
<span>                    destfile <span class="op">=</span> <span class="st">"kegg.rda"</span>, quiet <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/load.html" class="external-link">load</a></span><span class="op">(</span><span class="st">"kegg.rda"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.remove</a></span><span class="op">(</span><span class="st">"kegg.rda"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] TRUE</span></span></code></pre>
<p>Next we are performing the enrichment analysis with a p-value cut-off
of p-value &lt; 0.05</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ora.res</span> <span class="op">&lt;-</span> <span class="fu">spammR</span><span class="fu">::</span><span class="fu"><a href="../reference/enrich_ora.html">enrich_ora</a></span><span class="op">(</span><span class="va">diff_ex</span>, geneset <span class="op">=</span> <span class="va">keggPathwayToKo</span>, </span>
<span>                              geneset_name <span class="op">=</span> <span class="st">"KEGG pathways"</span>, </span>
<span>                              feature_column <span class="op">=</span> <span class="st">"KO"</span>, pval_thresh <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="va">filtered_res</span> <span class="op">&lt;-</span> <span class="va">ora.res</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="va">.</span>, var <span class="op">=</span> <span class="st">'KEGG_Pathway'</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">oddsratio</span> <span class="op">&gt;</span> <span class="fl">1</span> <span class="op">&amp;</span> <span class="va">BH_pvalue</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="op">-</span><span class="va">oddsratio</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">KEGG_Pathway</span>, <span class="va">oddsratio</span>, <span class="va">BH_pvalue</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pathway_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">keggPathwayToKo</span><span class="op">$</span><span class="va">name</span>, <span class="va">keggPathwayToKo</span><span class="op">$</span><span class="va">desc</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/rename.html" class="external-link">rename</a></span><span class="op">(</span><span class="va">.</span>, KEGG_Pathway <span class="op">=</span> <span class="va">V1</span>, KEGG_pathway_name <span class="op">=</span> <span class="va">V2</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: The `x` argument of `as_tibble.matrix()` must have unique column names if</span></span>
<span><span class="co">## `.name_repair` is omitted as of tibble 2.0.0.</span></span>
<span><span class="co">## <span style="color: #00BBBB;">ℹ</span> Using compatibility `.name_repair`.</span></span>
<span><span class="co">## <span style="color: #555555;">This warning is displayed once per session.</span></span></span>
<span><span class="co">## <span style="color: #555555;">Call `lifecycle::last_lifecycle_warnings()` to see where this warning was</span></span></span>
<span><span class="co">## <span style="color: #555555;">generated.</span></span></span></code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html" class="external-link">left_join</a></span><span class="op">(</span><span class="va">filtered_res</span>, <span class="va">pathway_names</span>, by <span class="op">=</span> <span class="st">'KEGG_Pathway'</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## <span style="color: #949494;"># A tibble: 8 × 4</span></span></span>
<span><span class="co">##   KEGG_Pathway oddsratio   BH_pvalue KEGG_pathway_name              </span></span>
<span><span class="co">##   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>            <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>                          </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">1</span> map01320          8.25 0.000<span style="text-decoration: underline;">266</span>    Sulfur cycle                   </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">2</span> map00540          7.64 0.000<span style="text-decoration: underline;">380</span>    Lipopolysaccharide biosynthesis</span></span>
<span><span class="co">## <span style="color: #BCBCBC;">3</span> map00480          5.00 0.030<span style="text-decoration: underline;">2</span>      Glutathione metabolism         </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">4</span> map00362          4.91 0.000<span style="text-decoration: underline;">011</span>8   Benzoate degradation           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">5</span> map00920          4.44 0.000<span style="text-decoration: underline;">266</span>    Sulfur metabolism              </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">6</span> map00627          3.98 0.018<span style="text-decoration: underline;">3</span>      Aminobenzoate degradation      </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">7</span> map02020          2.70 0.000<span style="text-decoration: underline;">000</span>387 Two-component system           </span></span>
<span><span class="co">## <span style="color: #BCBCBC;">8</span> map02010          2.25 0.000<span style="text-decoration: underline;">266</span>    ABC transporters</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ora.res</span> <span class="op">&lt;-</span> <span class="fu">tibble</span><span class="fu">::</span><span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html" class="external-link">rownames_to_column</a></span><span class="op">(</span><span class="va">ora.res</span>, <span class="st">"pathways"</span><span class="op">)</span></span>
<span><span class="va">pathway_prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rspatial.github.io/terra/reference/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">ora.res</span>, <span class="va">pathways</span> <span class="op">==</span> <span class="st">"map01320"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">ingroupnames</span><span class="op">)</span></span>
<span><span class="va">pathway_prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">pathway_prots</span><span class="op">[</span><span class="fl">1</span>, <span class="op">]</span>, split <span class="op">=</span> <span class="st">", "</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/spatial_heatmap.html">spatial_heatmap</a></span><span class="op">(</span></span>
<span>    <span class="va">microbiome.spe</span>,</span>
<span>    feature <span class="op">=</span> <span class="va">pathway_prots</span>,</span>
<span>    assay_name <span class="op">=</span> <span class="st">"KO"</span>,</span>
<span>    sample_id <span class="op">=</span> <span class="st">"1000 Soils"</span>,</span>
<span>    image_id <span class="op">=</span> <span class="st">"0"</span>,</span>
<span>    label_column <span class="op">=</span> <span class="st">"Desert"</span>,</span>
<span>    sample_label_size <span class="op">=</span> <span class="fl">2.0</span>,</span>
<span>    title_size <span class="op">=</span> <span class="fl">10</span>, </span>
<span>    metric_display <span class="op">=</span> <span class="st">"Enriched KO Groups"</span>,</span>
<span>    plot_title <span class="op">=</span> <span class="st">"Sulfur cycle"</span>,</span>
<span>    interactive <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: Removed 2 rows containing missing values or values outside the scale range</span></span>
<span><span class="co">## (`geom_label()`).</span></span></code></pre>
<p><img src="spatMicrobiome_files/figure-html/second%20heatmap-1.png" class="r-plt" alt="" width="700"></p>
</div>
<div class="section level3">
<h3 id="session-information">Session Information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h3>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.5.2 (2025-10-31)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Tahoe 26.2</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.5-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: America/Los_Angeles</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">## [8] base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] terra_1.8-93                sf_1.0-24                  </span></span>
<span><span class="co">##  [3] dplyr_1.1.4                 magrittr_2.0.4             </span></span>
<span><span class="co">##  [5] tibble_3.3.1                ggplot2_4.0.1              </span></span>
<span><span class="co">##  [7] spammR_0.99.17              limma_3.66.0               </span></span>
<span><span class="co">##  [9] SpatialExperiment_1.20.0    SingleCellExperiment_1.32.0</span></span>
<span><span class="co">## [11] SummarizedExperiment_1.40.0 Biobase_2.70.0             </span></span>
<span><span class="co">## [13] GenomicRanges_1.62.1        Seqinfo_1.0.0              </span></span>
<span><span class="co">## [15] IRanges_2.44.0              S4Vectors_0.48.0           </span></span>
<span><span class="co">## [17] BiocGenerics_0.56.0         generics_0.1.4             </span></span>
<span><span class="co">## [19] MatrixGenerics_1.22.0       matrixStats_1.5.0          </span></span>
<span><span class="co">## [21] BiocStyle_2.38.0           </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] DBI_1.2.3           deldir_2.0-4        gridExtra_2.3      </span></span>
<span><span class="co">##   [4] s2_1.1.9            rlang_1.1.7         otel_0.2.0         </span></span>
<span><span class="co">##   [7] e1071_1.7-17        compiler_4.5.2      png_0.1-8          </span></span>
<span><span class="co">##  [10] systemfonts_1.3.1   vctrs_0.7.1         wk_0.9.5           </span></span>
<span><span class="co">##  [13] pkgconfig_2.0.3     fastmap_1.2.0       backports_1.5.0    </span></span>
<span><span class="co">##  [16] magick_2.9.0        XVector_0.50.0      labeling_0.4.3     </span></span>
<span><span class="co">##  [19] ggraph_2.2.2        utf8_1.2.6          rmarkdown_2.30     </span></span>
<span><span class="co">##  [22] tzdb_0.5.0          ragg_1.5.0          purrr_1.2.1        </span></span>
<span><span class="co">##  [25] xfun_0.56           cachem_1.1.0        jsonlite_2.0.0     </span></span>
<span><span class="co">##  [28] DelayedArray_0.36.0 tweenr_2.0.3        broom_1.0.12       </span></span>
<span><span class="co">##  [31] R6_2.6.1            bslib_0.10.0        RColorBrewer_1.1-3 </span></span>
<span><span class="co">##  [34] reticulate_1.44.1   boot_1.3-32         car_3.1-3          </span></span>
<span><span class="co">##  [37] jquerylib_0.1.4     leapR_0.99.6        Rcpp_1.1.1         </span></span>
<span><span class="co">##  [40] bookdown_0.46       knitr_1.51          readr_2.1.6        </span></span>
<span><span class="co">##  [43] Matrix_1.7-4        igraph_2.2.1        tidyselect_1.2.1   </span></span>
<span><span class="co">##  [46] rstudioapi_0.18.0   abind_1.4-8         yaml_2.3.12        </span></span>
<span><span class="co">##  [49] viridis_0.6.5       codetools_0.2-20    lattice_0.22-7     </span></span>
<span><span class="co">##  [52] withr_3.0.2         S7_0.2.1            evaluate_1.0.5     </span></span>
<span><span class="co">##  [55] desc_1.4.3          units_1.0-0         proxy_0.4-29       </span></span>
<span><span class="co">##  [58] spData_2.3.4        polyclip_1.10-7     pillar_1.11.1      </span></span>
<span><span class="co">##  [61] BiocManager_1.30.27 ggpubr_0.6.2        carData_3.0-5      </span></span>
<span><span class="co">##  [64] KernSmooth_2.23-26  plotly_4.12.0       sp_2.2-0           </span></span>
<span><span class="co">##  [67] hms_1.1.4           scales_1.4.0        class_7.3-23       </span></span>
<span><span class="co">##  [70] glue_1.8.0          lazyeval_0.2.2      tools_4.5.2        </span></span>
<span><span class="co">##  [73] data.table_1.18.2.1 ggnewscale_0.5.2    ggsignif_0.6.4     </span></span>
<span><span class="co">##  [76] fs_1.6.6            graphlayouts_1.2.2  tidygraph_1.3.1    </span></span>
<span><span class="co">##  [79] grid_4.5.2          spdep_1.4-1         impute_1.84.0      </span></span>
<span><span class="co">##  [82] tidyr_1.3.2         ggforce_0.5.0       Formula_1.2-5      </span></span>
<span><span class="co">##  [85] cli_3.6.5           textshaping_1.0.4   S4Arrays_1.10.1    </span></span>
<span><span class="co">##  [88] viridisLite_0.4.2   gtable_0.3.6        rstatix_0.7.3      </span></span>
<span><span class="co">##  [91] sass_0.4.10         digest_0.6.39       classInt_0.4-11    </span></span>
<span><span class="co">##  [94] SparseArray_1.10.8  ggrepel_0.9.6       rjson_0.2.23       </span></span>
<span><span class="co">##  [97] htmlwidgets_1.6.4   farver_2.1.2        memoise_2.0.1      </span></span>
<span><span class="co">## [100] htmltools_0.5.9     pkgdown_2.2.0       lifecycle_1.0.5    </span></span>
<span><span class="co">## [103] httr_1.4.7          statmod_1.5.1       MASS_7.3-65</span></span></code></pre>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sara Gosline.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer>
</div>





  </body>
</html>
