<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Calculate functional or pathway enrichment — enrich_ora • spammR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Calculate functional or pathway enrichment — enrich_ora"><meta name="description" content="enrich_ora() calculates over-representation statistics
(ORA) using an interest list of genes from differential
expression results in spammR and gene sets (either the ones provided in
spammR or user supplied)
This function uses results from calc_spatial_diff_ex. Interest list of
genes for ORA is obtained from spatialDiffEx results
based on the criteria specified in this function.It is a wrapper for the
leapR package which is required
For ORA using an external or already defined interest list of genes and
gene sets, use leapR functions directly"><meta property="og:description" content="enrich_ora() calculates over-representation statistics
(ORA) using an interest list of genes from differential
expression results in spammR and gene sets (either the ones provided in
spammR or user supplied)
This function uses results from calc_spatial_diff_ex. Interest list of
genes for ORA is obtained from spatialDiffEx results
based on the criteria specified in this function.It is a wrapper for the
leapR package which is required
For ORA using an external or already defined interest list of genes and
gene sets, use leapR functions directly"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">spammR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.99.17</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/lipidProt.html">spammR lipidomic and proteomic integration</a></li>
    <li><a class="dropdown-item" href="../articles/spatMicrobiome.html">spammR microbiome evaluation</a></li>
    <li><a class="dropdown-item" href="../articles/spatProt.html">spammR Spatial Proteomics Example</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Calculate functional or pathway enrichment</h1>

      <div class="d-none name"><code>enrich_ora.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>enrich_ora()</code> calculates over-representation statistics
(ORA) using an interest list of genes from differential
expression results in spammR and gene sets (either the ones provided in
spammR or user supplied)
This function uses results from <code>calc_spatial_diff_ex</code>. Interest list of
genes for ORA is obtained from spatialDiffEx results
based on the criteria specified in this function.It is a wrapper for the
<code>leapR</code> package which is required
For ORA using an external or already defined interest list of genes and
gene sets, use leapR functions directly</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">enrich_ora</span><span class="op">(</span></span>
<span>  <span class="va">spe</span>,</span>
<span>  <span class="va">geneset</span>,</span>
<span>  <span class="va">feature_column</span>,</span>
<span>  pval_type_forThresh <span class="op">=</span> <span class="st">"adjusted_pval"</span>,</span>
<span>  pval_thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  logFC_lowerThresh <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  logFC_upperThresh <span class="op">=</span> <span class="cn">NA</span>,</span>
<span>  geneset_name <span class="op">=</span> <span class="st">"msigdb"</span>,</span>
<span>  <span class="va">sortResultsBy</span>,</span>
<span>  comparison_name <span class="op">=</span> <span class="st">""</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-spe">spe<a class="anchor" aria-label="anchor" href="#arg-spe"></a></dt>
<dd><p>SpatialExperiment object containing spatial omics data and
spatial diffex results</p></dd>


<dt id="arg-geneset">geneset<a class="anchor" aria-label="anchor" href="#arg-geneset"></a></dt>
<dd><p>in GMT format</p></dd>


<dt id="arg-feature-column">feature_column<a class="anchor" aria-label="anchor" href="#arg-feature-column"></a></dt>
<dd><p>Column of rowData that maps to gene set</p></dd>


<dt id="arg-pval-type-forthresh">pval_type_forThresh<a class="anchor" aria-label="anchor" href="#arg-pval-type-forthresh"></a></dt>
<dd><p>Choose from "adjusted_pval" or "pval". Type of
p-value that should be used for
filtering statistically significant results. Default is adjusted p-value
for multiple hypotheses correction.</p></dd>


<dt id="arg-pval-thresh">pval_thresh<a class="anchor" aria-label="anchor" href="#arg-pval-thresh"></a></dt>
<dd><p>value to use for filtering based on
pval_type_forThreshold. Default is 0.05.
Values less than pval_thresh will be kept.</p></dd>


<dt id="arg-logfc-lowerthresh">logFC_lowerThresh<a class="anchor" aria-label="anchor" href="#arg-logfc-lowerthresh"></a></dt>
<dd><p>Lower threshold for log Fold Change, to be used
for filtering spatialDiffEx results. Default is NA</p></dd>


<dt id="arg-logfc-upperthresh">logFC_upperThresh<a class="anchor" aria-label="anchor" href="#arg-logfc-upperthresh"></a></dt>
<dd><p>Upper threshold for log Fold Change, to be used
for filtering spatialDiffEx results. Default is NA</p></dd>


<dt id="arg-geneset-name">geneset_name<a class="anchor" aria-label="anchor" href="#arg-geneset-name"></a></dt>
<dd><p>Name of geneset provided</p></dd>


<dt id="arg-sortresultsby">sortResultsBy<a class="anchor" aria-label="anchor" href="#arg-sortresultsby"></a></dt>
<dd><p>For sorting ORA results, choose from the following
column names: "BH_pvalue" (default)</p></dd>


<dt id="arg-comparison-name">comparison_name<a class="anchor" aria-label="anchor" href="#arg-comparison-name"></a></dt>
<dd><p>Example: "RSPv_vs_others" Text to indicate in
results data frame, which spatial groups were
compared for the interest list of genes</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A dataframe containing results from over-representation
analysis of members of gene sets in the
interest list of genes based on filtering criteria above.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">smallPancData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">pancMeta</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">protMeta</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">pooledPanc</span> <span class="op">&lt;-</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind_cols.html" class="external-link">bind_cols</a></span><span class="op">(</span><span class="va">smallPancData</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">panc.spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="convert_to_spe.html">convert_to_spe</a></span><span class="op">(</span><span class="va">pooledPanc</span>, <span class="va">pancMeta</span>, <span class="va">protMeta</span>, </span></span>
<span class="r-in"><span>feature_meta_colname <span class="op">=</span> <span class="st">"pancProts"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Spatial object created without spatial coordinate </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>          column names provided. Distance based analysis will not be enabled.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Note: Only mapping metadata for 2986 features out of 3000 data points</span>
<span class="r-in"><span><span class="va">diffex.spe</span> <span class="op">&lt;-</span> <span class="fu"><a href="calc_spatial_diff_ex.html">calc_spatial_diff_ex</a></span><span class="op">(</span><span class="va">panc.spe</span>, category_col <span class="op">=</span> <span class="st">"IsletOrNot"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Partial NA coefficients for 2 probe(s)</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> We found 0 features with a logFC greater than 1 and </span>
<span class="r-msg co"><span class="r-pr">#&gt;</span>                  an ajusted p-value less than 0.05</span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">leapR</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"krbpaths"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">ora.res</span> <span class="op">&lt;-</span> <span class="fu">enrich_ora</span><span class="op">(</span><span class="va">diffex.spe</span>, geneset <span class="op">=</span> <span class="va">krbpaths</span>, </span></span>
<span class="r-in"><span>         geneset_name <span class="op">=</span> <span class="st">"krbpaths"</span>, feature_column <span class="op">=</span> <span class="st">"PrimaryGeneName"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Sara Gosline.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

